diff --git a/dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/exchange/support/header/ReconnectTimerTask.java b/dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/exchange/support/header/ReconnectTimerTask.java
index 93c790330e7..2b7dca552c8 100644
--- a/dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/exchange/support/header/ReconnectTimerTask.java
+++ b/dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/exchange/support/header/ReconnectTimerTask.java
@@ -47,7 +47,7 @@ protected void doTask(Channel channel) {
                             + heartbeatTimeout + "ms");
                     ((Client) channel).reconnect();
                 } catch (Throwable t) {
-                    logger.warn("Exception when reconnect to remote channel " + channel.getRemoteAddress(), t);
+                    // do nothing
                 }
             } else {
                 try {
