diff --git a/resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala b/resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala
index 137cf69d22a55..64698b55c6bb6 100644
--- a/resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala
+++ b/resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala
@@ -52,6 +52,7 @@ private[mesos] class MesosClusterDispatcher(
   extends Logging {
 
   {
+    // This doesn't support authentication because the RestSubmissionServer doesn't support it.
     val authKey = SecurityManager.SPARK_AUTH_SECRET_CONF
     require(conf.getOption(authKey).isEmpty,
       s"The MesosClusterDispatcher does not support authentication via ${authKey}.  It is not " +
