diff --git a/core/src/test/scala/org/apache/spark/scheduler/BlacklistIntegrationSuite.scala b/core/src/test/scala/org/apache/spark/scheduler/BlacklistIntegrationSuite.scala
index 1b80f6211cf1f..1100156bb5cc3 100644
--- a/core/src/test/scala/org/apache/spark/scheduler/BlacklistIntegrationSuite.scala
+++ b/core/src/test/scala/org/apache/spark/scheduler/BlacklistIntegrationSuite.scala
@@ -42,7 +42,7 @@ class BlacklistIntegrationSuite extends SchedulerIntegrationSuite[MultiExecutorM
 
   override def test(name: String, testTags: org.scalatest.Tag*)(body: => Unit): Unit = {
     var lastThreads = Set[Long]()
-    (0 until 5000).foreach { idx =>
+    (0 until 500).foreach { idx =>
       super.test(s"$name: $idx", testTags: _*) {
         body
       }
diff --git a/core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala b/core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala
index 04b84f27727eb..050808171fdce 100644
--- a/core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala
+++ b/core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala
@@ -328,7 +328,7 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with Timeou
 
   override def test(name: String, testTags: org.scalatest.Tag*)(body: => Unit): Unit = {
     var lastThreads = Set[Long]()
-    (0 until 1000).foreach { idx =>
+    (0 until 50).foreach { idx =>
       super.test(s"$name: $idx", testTags: _*) {
         body
       }
