diff --git a/iocore/eventsystem/I_Continuation.h b/iocore/eventsystem/I_Continuation.h
index a6c4cd70556..ec7c3f014ff 100644
--- a/iocore/eventsystem/I_Continuation.h
+++ b/iocore/eventsystem/I_Continuation.h
@@ -109,7 +109,7 @@ class Continuation : private force_VFPT_to_top
 #endif
 
   /**
-    The Contination's lock.
+    The Continuation's lock.
 
     A reference counted pointer to the Continuation's lock. This
     lock is initialized in the constructor and should not be set
