diff --git a/examples/camel-example-debezium/pom.xml b/examples/camel-example-debezium/pom.xml
index dda4aaf2f97e..393075115412 100644
--- a/examples/camel-example-debezium/pom.xml
+++ b/examples/camel-example-debezium/pom.xml
@@ -33,7 +33,7 @@
     <description>An example for Debezium Component</description>
 
     <properties>
-        <category>Change Data Capture</category>
+        <category>Database</category>
     </properties>
 
     <dependencies>
diff --git a/examples/camel-example-debezium/src/main/java/org/apache/camel/example/debezium/DebeziumMySqlConsumerToKinesis.java b/examples/camel-example-debezium/src/main/java/org/apache/camel/example/debezium/DebeziumMySqlConsumerToKinesis.java
index 2e9513fdfa7d..328881528ce4 100644
--- a/examples/camel-example-debezium/src/main/java/org/apache/camel/example/debezium/DebeziumMySqlConsumerToKinesis.java
+++ b/examples/camel-example-debezium/src/main/java/org/apache/camel/example/debezium/DebeziumMySqlConsumerToKinesis.java
@@ -42,7 +42,7 @@ private DebeziumMySqlConsumerToKinesis() {
 
     public static void main(String[] args) throws Exception {
 
-        LOG.info("About to run Debezium integration...");
+        LOG.debug("About to run Debezium integration...");
 
         final CamelContext camelContext = new DefaultCamelContext();
 
diff --git a/examples/camel-example-debezium/src/main/java/org/apache/camel/example/debezium/KinesisProducerToCassandra.java b/examples/camel-example-debezium/src/main/java/org/apache/camel/example/debezium/KinesisProducerToCassandra.java
index d0b85be9b6f6..007b598a64b7 100644
--- a/examples/camel-example-debezium/src/main/java/org/apache/camel/example/debezium/KinesisProducerToCassandra.java
+++ b/examples/camel-example-debezium/src/main/java/org/apache/camel/example/debezium/KinesisProducerToCassandra.java
@@ -40,7 +40,7 @@ private KinesisProducerToCassandra() {
 
     public static void main(String[] args) throws Exception {
 
-        LOG.info("About to run Kinesis to Cassandra integration...");
+        LOG.debug("About to run Kinesis to Cassandra integration...");
 
         final CamelContext camelContext = new DefaultCamelContext();
 
diff --git a/examples/pom.xml b/examples/pom.xml
index b9b8b04b7b9c..53254b2ebe5c 100644
--- a/examples/pom.xml
+++ b/examples/pom.xml
@@ -67,6 +67,7 @@
         <module>camel-example-cxf-tomcat</module>
         <module>camel-example-cxf-ws-security-signature</module>
         <module>camel-example-console</module>
+        <module>camel-example-debezium</module>
         <module>camel-example-ehcache-blueprint</module>
         <module>camel-example-ftp</module>
         <module>camel-example-google-pubsub</module>
