diff --git a/airflow/__init__.py b/airflow/__init__.py
index 1ed188cc45886..7da6ea1b47738 100644
--- a/airflow/__init__.py
+++ b/airflow/__init__.py
@@ -20,7 +20,7 @@
 `airflow.www.login`
 """
 from builtins import object
-__version__ = "1.6.2"
+__version__ = "1.7.0"
 
 import logging
 import os
diff --git a/setup.py b/setup.py
index d8306d436c711..924aa3245c394 100644
--- a/setup.py
+++ b/setup.py
@@ -5,7 +5,7 @@
 import sys
 
 # Kept manually in sync with airflow.__version__
-version = '1.6.2'
+version = '1.7.0'
 
 
 class Tox(TestCommand):
@@ -104,6 +104,7 @@ def run(self):
 setup(
     name='airflow',
     description='Programmatically author, schedule and monitor data pipelines',
+    license='Apache License 2.0',
     version=version,
     packages=find_packages(),
     package_data={'': ['airflow/alembic.ini']},
@@ -166,6 +167,17 @@ def run(self):
         'github_enterprise': github_enterprise,
         'qds': qds
     },
+    classifiers={
+        'Development Status :: 5 - Production/Stable',
+        'Environment :: Console',
+        'Environment :: Web Environment',
+        'Intended Audience :: Developers',
+        'Intended Audience :: System Administrators',
+        'License :: OSI Approved :: Apache Software License',
+        'Programming Language :: Python :: 2.7',
+        'Programming Language :: Python :: 3.4',
+        'Topic :: System :: Monitoring',
+    },
     author='Maxime Beauchemin',
     author_email='maximebeauchemin@gmail.com',
     url='https://github.com/airbnb/airflow',
