diff --git a/tests/python/frontend/darknet/test_forward.py b/tests/python/frontend/darknet/test_forward.py
index 4d45307cb4c7..3545e8a902bd 100644
--- a/tests/python/frontend/darknet/test_forward.py
+++ b/tests/python/frontend/darknet/test_forward.py
@@ -16,7 +16,7 @@
 # under the License.
 """
 Test Darknet Models
-=====================
+===================
 This article is a test script to test darknet models with Relay.
 All the required models and libraries will be downloaded from the internet
 by the script.
@@ -31,13 +31,13 @@
 from tvm.relay.frontend.darknet import ACTIVATION
 from tvm import relay
 
+REPO_URL = 'https://github.com/dmlc/web-data/blob/master/darknet/'
 DARKNET_LIB = 'libdarknet2.0.so'
-DARKNETLIB_URL = 'https://github.com/siju-samuel/darknet/blob/master/lib/' \
-                                    + DARKNET_LIB + '?raw=true'
+DARKNETLIB_URL = REPO_URL + 'lib/' + DARKNET_LIB + '?raw=true'
 LIB = __darknetffi__.dlopen(download_testdata(DARKNETLIB_URL, DARKNET_LIB, module='darknet'))
 
 DARKNET_TEST_IMAGE_NAME = 'dog.jpg'
-DARKNET_TEST_IMAGE_URL = 'https://github.com/siju-samuel/darknet/blob/master/data/' + DARKNET_TEST_IMAGE_NAME +'?raw=true'
+DARKNET_TEST_IMAGE_URL = REPO_URL + 'data/' + DARKNET_TEST_IMAGE_NAME +'?raw=true'
 DARKNET_TEST_IMAGE_PATH = download_testdata(DARKNET_TEST_IMAGE_URL, DARKNET_TEST_IMAGE_NAME, module='data')
 
 def _read_memory_buffer(shape, data, dtype='float32'):
diff --git a/tutorials/frontend/from_darknet.py b/tutorials/frontend/from_darknet.py
index bc23b9bddca0..2658a353e34e 100644
--- a/tutorials/frontend/from_darknet.py
+++ b/tutorials/frontend/from_darknet.py
@@ -16,7 +16,7 @@
 # under the License.
 """
 Compile YOLO-V2 and YOLO-V3 in DarkNet Models
-=================================
+=============================================
 **Author**: `Siju Samuel <https://siju-samuel.github.io/>`_
 
 This article is an introductory tutorial to deploy darknet models with TVM.
@@ -54,7 +54,7 @@
 # Download cfg and weights file if first time.
 CFG_NAME = MODEL_NAME + '.cfg'
 WEIGHTS_NAME = MODEL_NAME + '.weights'
-REPO_URL = 'https://github.com/siju-samuel/darknet/blob/master/'
+REPO_URL = 'https://github.com/dmlc/web-data/blob/master/darknet/'
 CFG_URL = REPO_URL + 'cfg/' + CFG_NAME + '?raw=true'
 WEIGHTS_URL = 'https://pjreddie.com/media/files/' + WEIGHTS_NAME
 
@@ -100,11 +100,10 @@
 [neth, netw] = shape['data'][2:] # Current image shape is 608x608
 ######################################################################
 # Load a test image
-# --------------------------------------------------------------------
+# -----------------
 test_image = 'dog.jpg'
 print("Loading the test image...")
-img_url = 'https://github.com/siju-samuel/darknet/blob/master/data/' + \
-          test_image + '?raw=true'
+img_url = REPO_URL + 'data/' + test_image + '?raw=true'
 img_path = download_testdata(img_url, test_image, "data")
 
 data = tvm.relay.testing.darknet.load_image(img_path, netw, neth)
@@ -164,9 +163,9 @@
 tvm.relay.testing.yolo_detection.do_nms_sort(dets, last_layer.classes, nms_thresh)
 
 coco_name = 'coco.names'
-coco_url = 'https://github.com/siju-samuel/darknet/blob/master/data/' + coco_name + '?raw=true'
+coco_url = REPO_URL + 'data/' + coco_name + '?raw=true'
 font_name = 'arial.ttf'
-font_url = 'https://github.com/siju-samuel/darknet/blob/master/data/' + font_name + '?raw=true'
+font_url = REPO_URL + 'data/' + font_name + '?raw=true'
 coco_path = download_testdata(coco_url, coco_name, module='data')
 font_path = download_testdata(font_url, font_name, module='data')
 
