diff --git a/core/src/main/scala/org/apache/spark/storage/BlockManager.scala b/core/src/main/scala/org/apache/spark/storage/BlockManager.scala
index 58b7acfd95831..e0478ad09601d 100644
--- a/core/src/main/scala/org/apache/spark/storage/BlockManager.scala
+++ b/core/src/main/scala/org/apache/spark/storage/BlockManager.scala
@@ -1912,7 +1912,11 @@ private[spark] class BlockManager(
 
     private val blockReplicationThread = new Thread {
       override def run(): Unit = {
-        while (blockManagerDecommissioning && !stopped) {
+        var failures = 0
+        while (blockManagerDecommissioning
+          && !stopped
+          && !Thread.interrupted()
+          && failures < 20) {
           try {
             logDebug("Attempting to replicate all cached RDD blocks")
             decommissionRddCacheBlocks()
@@ -1923,8 +1927,9 @@ private[spark] class BlockManager(
               logInfo("Interrupted during migration, will not refresh migrations.")
               stopped = true
             case NonFatal(e) =>
-              logError("Error occurred while trying to " +
-                "replicate cached RDD blocks for block manager decommissioning", e)
+              failures += 1
+              logError("Error occurred while trying to replicate cached RDD blocks" +
+                s" for block manager decommissioning (failure count: $failures)", e)
           }
         }
       }
