diff --git a/tests/python/unittest/test_gluon.py b/tests/python/unittest/test_gluon.py
index a6932d252dd2..e8ef704c4ec2 100644
--- a/tests/python/unittest/test_gluon.py
+++ b/tests/python/unittest/test_gluon.py
@@ -2007,6 +2007,7 @@ def hybrid_forward(self, F, x):
 
 
 @with_seed()
+@unittest.skip('Flaky test: https://github.com/apache/incubator-mxnet/issues/12767')
 def test_slice_batchnorm_reshape_batchnorm():
     class Net(gluon.HybridBlock):
         def __init__(self, shape, slice, **kwargs):
