diff --git a/external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala b/external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala
index 4d928bd0afd21..f124850bc6ce3 100644
--- a/external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala
+++ b/external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala
@@ -138,14 +138,13 @@ class KafkaTestUtils(
     kdc = new MiniKdc(kdcConf, kdcDir)
     kdc.start()
     // TODO https://issues.apache.org/jira/browse/SPARK-30037
-    //  Need to customize krb5.conf like Kafka
+    // Need to build spark's own MiniKDC and customize krb5.conf like Kafka
     rewriteKrb5Conf()
     kdcReady = true
   }
 
   /**
-   * In this method we rewrite krb5.conf to make kdc nad client use the same enctypes
-   *
+   * In this method we rewrite krb5.conf to make kdc and client use the same enctypes
    */
   private def rewriteKrb5Conf(): Unit = {
     val krb5Conf = Source.fromFile(kdc.getKrb5conf, "UTF-8").getLines()
