diff --git a/core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala b/core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala
index ebc843d9758ee..02e1bc2d62513 100644
--- a/core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala
+++ b/core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala
@@ -899,9 +899,11 @@ private[spark] class SparkSubmit extends Logging {
 
     try {
       app.start(childArgs.toArray, sparkConf)
+      System.exit(0)
     } catch {
       case t: Throwable =>
-        throw findCause(t)
+        logError("Error running job", t)
+        System.exit(-1)
     }
   }
 
