diff --git a/modules/core/src/main/java/org/apache/ignite/cluster/ClusterMetrics.java b/modules/core/src/main/java/org/apache/ignite/cluster/ClusterMetrics.java
index 87b719873694..751e62873d3f 100644
--- a/modules/core/src/main/java/org/apache/ignite/cluster/ClusterMetrics.java
+++ b/modules/core/src/main/java/org/apache/ignite/cluster/ClusterMetrics.java
@@ -696,9 +696,9 @@ public interface ClusterMetrics {
     public int getTotalNodes();
 
     /**
-     * Gets execution time for current partition map exchange.
+     * Gets execution duration for current partition map exchange in milliseconds.
      *
-     * @return Gets execution time for current partition map exchange. {@code 0} if there is no running PME.
+     * @return Gets execution duration for current partition map exchange in milliseconds. {@code 0} if there is no running PME.
      */
-    public long getCurrentPmeTime();
+    public long getCurrentPmeDuration();
 }
diff --git a/modules/core/src/main/java/org/apache/ignite/internal/ClusterLocalNodeMetricsMXBeanImpl.java b/modules/core/src/main/java/org/apache/ignite/internal/ClusterLocalNodeMetricsMXBeanImpl.java
index 9b9815a8032b..c3bcf803f70a 100644
--- a/modules/core/src/main/java/org/apache/ignite/internal/ClusterLocalNodeMetricsMXBeanImpl.java
+++ b/modules/core/src/main/java/org/apache/ignite/internal/ClusterLocalNodeMetricsMXBeanImpl.java
@@ -335,8 +335,8 @@ public ClusterLocalNodeMetricsMXBeanImpl(GridDiscoveryManager discoMgr) {
     }
 
     /** {@inheritDoc} */
-    @Override public long getCurrentPmeTime() {
-        return node.metrics().getCurrentPmeTime();
+    @Override public long getCurrentPmeDuration() {
+        return node.metrics().getCurrentPmeDuration();
     }
 
     /** {@inheritDoc} */
diff --git a/modules/core/src/main/java/org/apache/ignite/internal/ClusterMetricsMXBeanImpl.java b/modules/core/src/main/java/org/apache/ignite/internal/ClusterMetricsMXBeanImpl.java
index b3868173254b..316ea71ca749 100644
--- a/modules/core/src/main/java/org/apache/ignite/internal/ClusterMetricsMXBeanImpl.java
+++ b/modules/core/src/main/java/org/apache/ignite/internal/ClusterMetricsMXBeanImpl.java
@@ -362,8 +362,8 @@ private ClusterMetrics metrics() {
     }
 
     /** {@inheritDoc} */
-    @Override public long getCurrentPmeTime() {
-        return metrics().getCurrentPmeTime();
+    @Override public long getCurrentPmeDuration() {
+        return metrics().getCurrentPmeDuration();
     }
 
     /** {@inheritDoc} */
diff --git a/modules/core/src/main/java/org/apache/ignite/internal/ClusterMetricsSnapshot.java b/modules/core/src/main/java/org/apache/ignite/internal/ClusterMetricsSnapshot.java
index cdac831960b9..29c1e7f91e68 100644
--- a/modules/core/src/main/java/org/apache/ignite/internal/ClusterMetricsSnapshot.java
+++ b/modules/core/src/main/java/org/apache/ignite/internal/ClusterMetricsSnapshot.java
@@ -258,7 +258,7 @@ public class ClusterMetricsSnapshot implements ClusterMetrics {
     private long totalJobsExecTime = -1;
 
     /** */
-    private long currentPmeTime = -1;
+    private long currentPmeDuration = -1;
 
     /**
      * Create empty snapshot.
@@ -333,7 +333,7 @@ public ClusterMetricsSnapshot(ClusterGroup p) {
         outMesQueueSize = 0;
         heapTotal = 0;
         totalNodes = nodes.size();
-        currentPmeTime = 0;
+        currentPmeDuration = 0;
 
         for (ClusterNode node : nodes) {
             ClusterMetrics m = node.metrics();
@@ -411,7 +411,7 @@ public ClusterMetricsSnapshot(ClusterGroup p) {
 
             avgLoad += m.getCurrentCpuLoad();
 
-            currentPmeTime = max(currentPmeTime, m.getCurrentPmeTime());
+            currentPmeDuration = max(currentPmeDuration, m.getCurrentPmeDuration());
         }
 
         curJobExecTime /= size;
@@ -968,8 +968,8 @@ public void setCurrentIdleTime(long curIdleTime) {
     }
 
     /** {@inheritDoc} */
-    @Override public long getCurrentPmeTime() {
-        return currentPmeTime;
+    @Override public long getCurrentPmeDuration() {
+        return currentPmeDuration;
     }
 
     /**
@@ -1210,10 +1210,10 @@ public void setTotalNodes(int totalNodes) {
     /**
      * Sets execution time for current partition map exchange.
      *
-     * @param currentPmeTime Execution time for current partition map exchange.
+     * @param currentPmeDuration Execution time for current partition map exchange.
      */
-    public void setCurrentPmeTime(long currentPmeTime) {
-        this.currentPmeTime = currentPmeTime;
+    public void setCurrentPmeDuration(long currentPmeDuration) {
+        this.currentPmeDuration = currentPmeDuration;
     }
 
     /**
@@ -1368,7 +1368,7 @@ public static int serialize(byte[] data, int off, ClusterMetrics metrics) {
         buf.putInt(metrics.getOutboundMessagesQueueSize());
         buf.putInt(metrics.getTotalNodes());
         buf.putLong(metrics.getTotalJobsExecutionTime());
-        buf.putLong(metrics.getCurrentPmeTime());
+        buf.putLong(metrics.getCurrentPmeDuration());
 
         assert !buf.hasRemaining() : "Invalid metrics size [expected=" + METRICS_SIZE + ", actual="
             + (buf.position() - off) + ']';
@@ -1452,9 +1452,9 @@ public static ClusterMetrics deserialize(byte[] data, int off) {
             metrics.setTotalJobsExecutionTime(0);
 
         if (buf.remaining() >= 8)
-            metrics.setCurrentPmeTime(buf.getLong());
+            metrics.setCurrentPmeDuration(buf.getLong());
         else
-            metrics.setCurrentPmeTime(0);
+            metrics.setCurrentPmeDuration(0);
 
         return metrics;
     }
diff --git a/modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java b/modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java
index 1bc62c58d532..5645ee0005e7 100644
--- a/modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java
+++ b/modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java
@@ -18,6 +18,7 @@
 package org.apache.ignite.internal.managers.discovery;
 
 import java.util.Collection;
+import java.util.concurrent.TimeUnit;
 import org.apache.ignite.cluster.ClusterMetrics;
 import org.apache.ignite.internal.GridKernalContext;
 import org.apache.ignite.internal.processors.cache.GridCacheAdapter;
@@ -352,10 +353,10 @@ public ClusterMetricsImpl(GridKernalContext ctx, GridLocalMetrics vmMetrics, lon
     }
 
     /** {@inheritDoc} */
-    @Override public long getCurrentPmeTime() {
+    @Override public long getCurrentPmeDuration() {
         return (ctx.cache().context().exchange().lastTopologyFuture() == null ||
             ctx.cache().context().exchange().lastTopologyFuture().isDone()) ? 0 :
-            (System.nanoTime() - ctx.cache().context().exchange().lastTopologyFuture().getStartTime());
+            TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - ctx.cache().context().exchange().lastTopologyFuture().getStartTime());
     }
 
     /**
diff --git a/modules/core/src/main/java/org/apache/ignite/mxbean/ClusterMetricsMXBean.java b/modules/core/src/main/java/org/apache/ignite/mxbean/ClusterMetricsMXBean.java
index eae55ee72f46..5dd5ebcfe9e7 100644
--- a/modules/core/src/main/java/org/apache/ignite/mxbean/ClusterMetricsMXBean.java
+++ b/modules/core/src/main/java/org/apache/ignite/mxbean/ClusterMetricsMXBean.java
@@ -258,8 +258,8 @@ public interface ClusterMetricsMXBean extends ClusterMetrics {
     public int getTotalNodes();
 
     /** {@inheritDoc} */
-    @Override @MXBeanDescription("Current PME time.")
-    public long getCurrentPmeTime();
+    @Override @MXBeanDescription("Current PME duration in milliseconds.")
+    public long getCurrentPmeDuration();
 
     /**
      * Get count of total baseline nodes.
diff --git a/modules/core/src/test/java/org/apache/ignite/spi/discovery/ClusterMetricsSnapshotSerializeSelfTest.java b/modules/core/src/test/java/org/apache/ignite/spi/discovery/ClusterMetricsSnapshotSerializeSelfTest.java
index 8c45efe692b4..6970f8e0e063 100644
--- a/modules/core/src/test/java/org/apache/ignite/spi/discovery/ClusterMetricsSnapshotSerializeSelfTest.java
+++ b/modules/core/src/test/java/org/apache/ignite/spi/discovery/ClusterMetricsSnapshotSerializeSelfTest.java
@@ -149,7 +149,7 @@ private ClusterMetrics createMetrics() {
         metrics.setHeapMemoryTotal(48);
         metrics.setTotalNodes(49);
         metrics.setTotalJobsExecutionTime(50);
-        metrics.setCurrentPmeTime(51);
+        metrics.setCurrentPmeDuration(51);
 
         return metrics;
     }
@@ -208,6 +208,6 @@ private boolean isMetricsEquals(ClusterMetrics obj, ClusterMetrics obj1) {
             obj.getHeapMemoryTotal() == obj1.getHeapMemoryTotal() &&
             obj.getTotalNodes() == obj1.getTotalNodes() &&
             obj.getTotalJobsExecutionTime() == obj1.getTotalJobsExecutionTime() &&
-            obj.getCurrentPmeTime() == obj1.getCurrentPmeTime();
+            obj.getCurrentPmeDuration() == obj1.getCurrentPmeDuration();
     }
 }
diff --git a/modules/platforms/dotnet/Apache.Ignite.Core.Tests/ApiParity/ClusterMetricsParityTest.cs b/modules/platforms/dotnet/Apache.Ignite.Core.Tests/ApiParity/ClusterMetricsParityTest.cs
index 8176d0c8503a..21160c16dbb0 100644
--- a/modules/platforms/dotnet/Apache.Ignite.Core.Tests/ApiParity/ClusterMetricsParityTest.cs
+++ b/modules/platforms/dotnet/Apache.Ignite.Core.Tests/ApiParity/ClusterMetricsParityTest.cs
@@ -30,7 +30,7 @@ public class ClusterMetricsParityTest
         {
             // IGNITE-7127
             "TotalJobsExecutionTime",
-            "CurrentPmeTime"
+            "CurrentPmeDuration"
         };
 
         /// <summary>
