diff --git a/.github/actions/to-integration-tests/Dockerfile b/.github/actions/to-integration-tests/Dockerfile
index e304598da1a..8294dd09d9c 100644
--- a/.github/actions/to-integration-tests/Dockerfile
+++ b/.github/actions/to-integration-tests/Dockerfile
@@ -18,12 +18,12 @@
 FROM golang:1.14.2-buster
 
 RUN apt-get update &&\
-	apt-get install -y --no-install-recommends git gcc postgresql-client-11 &&\
+	apt-get install -y --no-install-recommends git gcc postgresql-client-11 envsubst &&\
 	rm -rf /var/lib/apt/lists/*
 RUN go get -v bitbucket.org/liamstask/goose/cmd/goose github.com/lib/pq
 RUN mv $GOPATH/bin/goose /bin/
 
-COPY entrypoint.sh /entrypoint.sh
+COPY entrypoint.sh database.json traffic-ops-test.json dbconf.yml cdn.json /
 RUN chmod a+x /entrypoint.sh
 
 ENTRYPOINT /entrypoint.sh
diff --git a/.github/actions/to-integration-tests/cdn.json b/.github/actions/to-integration-tests/cdn.json
new file mode 100644
index 00000000000..efda921f663
--- /dev/null
+++ b/.github/actions/to-integration-tests/cdn.json
@@ -0,0 +1,51 @@
+{
+	"hypnotoad" : {
+		"listen" : [
+			"https://not-a-real-host.test:1?cert=$PWD/localhost.crt&key=$PWD/localhost.key&verify=0x00&ciphers=AES128-GCM-SHA256:HIGH:!RC4:!MD5:!aNULL:!EDH:!ED"
+		],
+		"user" : "trafops",
+		"group" : "trafops",
+		"heartbeat_timeout" : 20,
+		"pid_file" : "/var/run/traffic_ops.pid",
+		"workers" : 12
+	},
+	"use_ims": true,
+	"traffic_ops_golang" : {
+		"insecure": true,
+		"port" : "6443",
+		"log_location_error": "stderr",
+		"log_location_warning": "stderr",
+		"log_location_info": null,
+		"log_location_debug": null,
+		"log_location_event": null,
+		"max_db_connections": 20,
+		"db_conn_max_lifetime_seconds": 60,
+		"db_query_timeout_seconds": 20,
+		"supported_ds_metrics": [ "kbps", "tps_total", "tps_2xx", "tps_3xx", "tps_4xx", "tps_5xx" ]
+	},
+	"cors" : {
+		"access_control_allow_origin" : "*"
+	},
+	"to" : {
+		"base_url" : "https://localhost",
+		"email_from" : "no-reply@traffic-ops.test",
+		"no_account_found_msg" : "A Traffic Ops user account is required for access. Please contact your Traffic Ops user administrator."
+	},
+	"portal" : {
+		"base_url" : "https://not-a-real-host.test/#!/",
+		"email_from" : "no-reply@traffic-portal.test",
+		"pass_reset_path" : "user",
+		"user_register_path" : "user"
+	},
+	"secrets" : [
+		"blahblah"
+	],
+	"geniso" : {
+		"iso_root_path" : "/opt/traffic_ops/app/public"
+	},
+	"inactivity_timeout" : 60,
+	"smtp" : {
+		"enabled" : false
+	},
+	"InfluxEnabled": false
+}
diff --git a/.github/actions/to-integration-tests/database.json b/.github/actions/to-integration-tests/database.json
new file mode 100644
index 00000000000..8d08eb40d49
--- /dev/null
+++ b/.github/actions/to-integration-tests/database.json
@@ -0,0 +1,10 @@
+{
+	"description": "Local PostgreSQL database on port 5432",
+	"dbname": "traffic_ops",
+	"hostname": "postgres",
+	"user": "traffic_ops",
+	"password": "twelve",
+	"port": "5432",
+	"ssl": false,
+	"type": "Pg"
+}
diff --git a/.github/actions/to-integration-tests/dbconf.yml b/.github/actions/to-integration-tests/dbconf.yml
new file mode 100644
index 00000000000..f565670b38d
--- /dev/null
+++ b/.github/actions/to-integration-tests/dbconf.yml
@@ -0,0 +1,23 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+version: "1.0"
+name: dbconf.yml
+
+test:
+  driver: postgres
+  open: host=postgres port=5432 user=traffic_ops password=twelve dbname=traffic_ops sslmode=disable
+
diff --git a/.github/actions/to-integration-tests/entrypoint.sh b/.github/actions/to-integration-tests/entrypoint.sh
index eb3451efc74..44a56b38dfa 100644
--- a/.github/actions/to-integration-tests/entrypoint.sh
+++ b/.github/actions/to-integration-tests/entrypoint.sh
@@ -23,14 +23,7 @@ ln -s "$PWD" "$SRCDIR/trafficcontrol"
 
 cd "$SRCDIR/trafficcontrol/traffic_ops/app/db"
 
-echo 'version: "1.0"
-name: dbconf.yml
-
-test:
-  driver: postgres
-  open: host=postgres port=5432 user=traffic_ops password=twelve dbname=traffic_ops sslmode=disable
-
-' > dbconf.yml
+mv /dbconf.yml ./
 
 psql -d postgresql://traffic_ops:twelve@postgres:5432/traffic_ops < ./create_tables.sql >/dev/null
 goose --env=test --path="$PWD" up
@@ -97,123 +90,16 @@ A22D22wvfs7CE3cUz/8UnvLM3kbTTu1WbbBbrHjAV47sAHjW/ckTqeo=
 -----END RSA PRIVATE KEY-----
 " > localhost.key
 
-cat <<-EOF >cdn.conf
-{
-	"hypnotoad" : {
-		"listen" : [
-			"https://not-a-real-host.test:1?cert=$PWD/localhost.crt&key=$PWD/localhost.key&verify=0x00&ciphers=AES128-GCM-SHA256:HIGH:!RC4:!MD5:!aNULL:!EDH:!ED"
-		],
-		"user" : "trafops",
-		"group" : "trafops",
-		"heartbeat_timeout" : 20,
-		"pid_file" : "/var/run/traffic_ops.pid",
-		"workers" : 12
-	},
-	"use_ims": true,
-	"traffic_ops_golang" : {
-		"insecure": true,
-		"port" : "6443",
-		"log_location_error": "stderr",
-		"log_location_warning": "stderr",
-		"log_location_info": null,
-		"log_location_debug": null,
-		"log_location_event": null,
-		"max_db_connections": 20,
-		"db_conn_max_lifetime_seconds": 60,
-		"db_query_timeout_seconds": 20,
-		"supported_ds_metrics": [ "kbps", "tps_total", "tps_2xx", "tps_3xx", "tps_4xx", "tps_5xx" ]
-	},
-	"cors" : {
-		"access_control_allow_origin" : "*"
-	},
-	"to" : {
-		"base_url" : "https://localhost",
-		"email_from" : "no-reply@traffic-ops.test",
-		"no_account_found_msg" : "A Traffic Ops user account is required for access. Please contact your Traffic Ops user administrator."
-	},
-	"portal" : {
-		"base_url" : "https://not-a-real-host.test/#!/",
-		"email_from" : "no-reply@traffic-portal.test",
-		"pass_reset_path" : "user",
-		"user_register_path" : "user"
-	},
-	"secrets" : [
-		"blahblah"
-	],
-	"geniso" : {
-		"iso_root_path" : "/opt/traffic_ops/app/public"
-	},
-	"inactivity_timeout" : 60,
-	"smtp" : {
-		"enabled" : false
-	},
-	"InfluxEnabled": false
-}
-EOF
-
-cat <<-EOF >database.conf
-{
-	"description": "Local PostgreSQL database on port 5432",
-	"dbname": "traffic_ops",
-	"hostname": "postgres",
-	"user": "traffic_ops",
-	"password": "twelve",
-	"port": "5432",
-	"ssl": false,
-	"type": "Pg"
-}
-EOF
+envsubst </cdn.json >cdn.conf
+mv /database.json ./database.conf
 
 ./traffic_ops_golang --cfg ./cdn.conf --dbcfg ./database.conf >out.log 2>err.log &
 
 cd ../testing/api/v1
 
-makeCFG() {
-	cat <<-EOF >traffic-ops-test.conf
-	{
-		"default": {
-			"logLocations": {
-				"debug": "stderr",
-				"error": "stderr",
-				"event": "stderr",
-				"info": "stderr",
-				"warning": "stderr"
-			},
-			"session": {
-				"timeoutInSecs": 60
-			}
-		},
-		"trafficOps": {
-			"URL": "https://localhost:6443",
-			"password": "twelve",
-			"users": {
-				"disallowed": "disallowed",
-				"operations": "operations",
-				"admin": "admin",
-				"federation": "federation",
-				"portal": "portal",
-				"readOnly": "readOnly",
-				"extension": "extension"
-			}
-		},
-		"trafficOpsDB": {
-			"dbname": "traffic_ops",
-			"description": "Test database to_test",
-			"hostname": "postgres",
-			"password": "twelve",
-			"port": "5432",
-			"type": "Pg",
-			"user": "traffic_ops"
-		},
-		"noPerl": true,
-		"noISO": true
-	}
-	EOF
-}
-
 CODE="0"
 
-makeCFG
+cp /traffic-ops-test.json ./traffic-ops-test.conf
 /usr/local/go/bin/go test -v --cfg ./traffic-ops-test.conf
 if [ "$?" -gt "0" ]; then
 	CODE="1"
@@ -221,7 +107,7 @@ fi
 rm traffic-ops-test.conf
 
 cd ../v2
-makeCFG
+cp /traffic-ops-test.json ./traffic-ops-test.conf
  /usr/local/go/bin/go test -v --cfg ./traffic-ops-test.conf
 if [ "$?" -gt "0" ]; then
 	CODE="1"
@@ -229,7 +115,7 @@ fi
 rm traffic-ops-test.conf
 
 cd ../v3
-makeCFG
+cp /traffic-ops-test.json ./traffic-ops-test.conf
  /usr/local/go/bin/go test -v --cfg ./traffic-ops-test.conf
 if [ "$?" -gt "0" ]; then
 	CODE="1"
diff --git a/.github/actions/to-integration-tests/traffic-ops-test.json b/.github/actions/to-integration-tests/traffic-ops-test.json
new file mode 100644
index 00000000000..78c96eac95b
--- /dev/null
+++ b/.github/actions/to-integration-tests/traffic-ops-test.json
@@ -0,0 +1,38 @@
+{
+	"default": {
+		"logLocations": {
+			"debug": "stderr",
+			"error": "stderr",
+			"event": "stderr",
+			"info": "stderr",
+			"warning": "stderr"
+		},
+		"session": {
+			"timeoutInSecs": 60
+		}
+	},
+	"trafficOps": {
+		"URL": "https://localhost:6443",
+		"password": "twelve",
+		"users": {
+			"disallowed": "disallowed",
+			"operations": "operations",
+			"admin": "admin",
+			"federation": "federation",
+			"portal": "portal",
+			"readOnly": "readOnly",
+			"extension": "extension"
+		}
+	},
+	"trafficOpsDB": {
+		"dbname": "traffic_ops",
+		"description": "Test database to_test",
+		"hostname": "postgres",
+		"password": "twelve",
+		"port": "5432",
+		"type": "Pg",
+		"user": "traffic_ops"
+	},
+	"noPerl": true,
+	"noISO": true
+}
