diff --git a/csharp/test/Apache.Arrow.Benchmarks/ArrowReaderBenchmark.cs b/csharp/test/Apache.Arrow.Benchmarks/ArrowReaderBenchmark.cs
index 8fffa111d983..4e491a2a6b12 100644
--- a/csharp/test/Apache.Arrow.Benchmarks/ArrowReaderBenchmark.cs
+++ b/csharp/test/Apache.Arrow.Benchmarks/ArrowReaderBenchmark.cs
@@ -118,7 +118,7 @@ private static double SumAllNumbers(RecordBatch recordBatch)
                         break;
                     case ArrowTypeId.Decimal128:
                         Decimal128Array decimalArray = (Decimal128Array)array;
-                        sum += (double)Sum(decimalArray);
+                        sum += Sum(decimalArray);
                         break;
                 }
             }
