diff --git a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
index 568cba3a7f922..b095007a4e65d 100644
--- a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
+++ b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
@@ -1019,11 +1019,11 @@ object DateTimeUtils {
       case TRUNC_TO_DAY =>
         val offset = timeZone.getOffset(millis)
         millis += offset
-        millis - millis % (MILLIS_PER_SECOND * SECONDS_PER_DAY) - offset
+        millis - Math.floorMod(millis, MILLIS_PER_SECOND * SECONDS_PER_DAY) - offset
       case TRUNC_TO_HOUR =>
         val offset = timeZone.getOffset(millis)
         millis += offset
-        millis - millis % (60 * 60 * MILLIS_PER_SECOND) - offset
+        millis - Math.floorMod(millis, 60 * 60 * MILLIS_PER_SECOND) - offset
       case TRUNC_TO_MINUTE =>
         millis - Math.floorMod(millis, 60 * MILLIS_PER_SECOND)
       case TRUNC_TO_SECOND =>
diff --git a/sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala b/sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala
index 76a3e3f13b973..20d78f093bff8 100644
--- a/sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala
+++ b/sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala
@@ -730,7 +730,7 @@ class DateFunctionsSuite extends QueryTest with SharedSQLContext {
         Row(Timestamp.valueOf("2015-07-24 22:00:00"))))
   }
 
-  test("SPARK-30793: truncate timestamps before the epoch to seconds and minutes") {
+  test("SPARK-30793, SPARK-30857: truncate timestamps before the epoch") {
     def checkTrunc(level: String, expected: String): Unit = {
       val df = Seq("1961-04-12 00:01:02.345")
         .toDF()
@@ -741,5 +741,7 @@ class DateFunctionsSuite extends QueryTest with SharedSQLContext {
 
     checkTrunc("SECOND", "1961-04-12 00:01:02")
     checkTrunc("MINUTE", "1961-04-12 00:01:00")
+    checkTrunc("HOUR", "1961-04-12 00:00:00")
+    checkTrunc("DAY", "1961-04-12 00:00:00")
   }
 }
