diff --git a/parquet/src/main/java/org/apache/iceberg/parquet/RemoveIds.java b/parquet/src/main/java/org/apache/iceberg/parquet/RemoveIds.java
index b68b5c00d9d..0db22bbf7b7 100644
--- a/parquet/src/main/java/org/apache/iceberg/parquet/RemoveIds.java
+++ b/parquet/src/main/java/org/apache/iceberg/parquet/RemoveIds.java
@@ -27,11 +27,6 @@
 import org.apache.parquet.schema.Types;
 
 public class RemoveIds extends ParquetTypeVisitor<Type> {
-  private final String name;
-
-  public RemoveIds(String name) {
-    this.name = name;
-  }
 
   @Override
   public Type message(MessageType message, List<Type> fields) {
@@ -39,7 +34,7 @@ public Type message(MessageType message, List<Type> fields) {
     for (Type field : struct(message.asGroupType(), fields).asGroupType().getFields()) {
       builder.addField(field);
     }
-    return builder.named(name);
+    return builder.named(message.getName());
   }
 
   @Override
@@ -74,8 +69,8 @@ public Type primitive(PrimitiveType primitive) {
         .named(primitive.getName());
   }
 
-  public static MessageType removeIds(org.apache.iceberg.Schema schema, String name) {
-    return (MessageType) ParquetTypeVisitor.visit(ParquetSchemaUtil.convert(schema, name), new RemoveIds(name));
+  public static MessageType removeIds(MessageType type, String name) {
+    return (MessageType) ParquetTypeVisitor.visit(type, new RemoveIds());
   }
 
 }
diff --git a/parquet/src/test/java/org/apache/iceberg/parquet/TestParquetSchemaUtil.java b/parquet/src/test/java/org/apache/iceberg/parquet/TestParquetSchemaUtil.java
index 6d769a61a92..f72a5157231 100644
--- a/parquet/src/test/java/org/apache/iceberg/parquet/TestParquetSchemaUtil.java
+++ b/parquet/src/test/java/org/apache/iceberg/parquet/TestParquetSchemaUtil.java
@@ -88,7 +88,7 @@ public void testAssignIdsByNameMapping() {
     NameMapping nameMapping = MappingUtil.create(schema);
     MessageType messageTypeWithIds = ParquetSchemaUtil.convert(schema, "parquet_type");
     MessageType messageTypeWithIdsFromNameMapping = ParquetSchemaUtil
-        .applyNameMapping(RemoveIds.removeIds(schema, "parquet_type"), nameMapping);
+        .applyNameMapping(RemoveIds.removeIds(messageTypeWithIds, "parquet_type"), nameMapping);
 
     Assert.assertEquals(messageTypeWithIds, messageTypeWithIdsFromNameMapping);
   }
