diff --git a/tests/kafkatest/services/kafka/kafka.py b/tests/kafkatest/services/kafka/kafka.py
index 46a0ddf55c6..cb5018caf5c 100644
--- a/tests/kafkatest/services/kafka/kafka.py
+++ b/tests/kafkatest/services/kafka/kafka.py
@@ -47,6 +47,8 @@ class KafkaService(JmxMixin, Service):
     # Kafka log segments etc go here
     DATA_LOG_DIR = os.path.join(PERSISTENT_ROOT, "kafka-data-logs")
     CONFIG_FILE = os.path.join(PERSISTENT_ROOT, "kafka.properties")
+    # Kafka Authorizer
+    SIMPLE_AUTHORIZER = "kafka.security.auth.SimpleAclAuthorizer"
 
     logs = {
         "kafka_operational_logs_info": {
@@ -61,7 +63,7 @@ class KafkaService(JmxMixin, Service):
     }
 
     def __init__(self, context, num_nodes, zk, security_protocol=SecurityConfig.PLAINTEXT, interbroker_security_protocol=SecurityConfig.PLAINTEXT,
-                 sasl_mechanism=SecurityConfig.SASL_MECHANISM_GSSAPI, use_authorizer=False, topics=None, version=TRUNK, quota_config=None, jmx_object_names=None,
+                 sasl_mechanism=SecurityConfig.SASL_MECHANISM_GSSAPI, authorizer_class_name=None, topics=None, version=TRUNK, quota_config=None, jmx_object_names=None,
                  jmx_attributes=[], zk_connect_timeout=5000):
         """
         :type context
@@ -79,8 +81,7 @@ def __init__(self, context, num_nodes, zk, security_protocol=SecurityConfig.PLAI
         self.sasl_mechanism = sasl_mechanism
         self.topics = topics
         self.minikdc = None
-        if use_authorizer:
-            self.authorizer_class_name = "kafka.security.auth.SimpleAclAuthorizer"
+        self.authorizer_class_name = authorizer_class_name
         #
         # In a heavily loaded and not very fast machine, it is
         # sometimes necessary to give more time for the zk client
diff --git a/tests/kafkatest/tests/security_rolling_upgrade_test.py b/tests/kafkatest/tests/security_rolling_upgrade_test.py
index 2b17eca6e57..f19d6b8e8f7 100644
--- a/tests/kafkatest/tests/security_rolling_upgrade_test.py
+++ b/tests/kafkatest/tests/security_rolling_upgrade_test.py
@@ -74,7 +74,7 @@ def roll_in_secured_settings(self, client_protocol, broker_protocol):
 
         # Roll cluster to disable PLAINTEXT port
         self.kafka.close_port('PLAINTEXT')
-        self.kafka.authorizer_class_name = "kafka.security.auth.SimpleAclAuthorizer"
+        self.kafka.authorizer_class_name = KafkaService.SIMPLE_AUTHORIZER
         self.acls.set_acls(client_protocol, self.kafka, self.zk, self.topic, self.group)
         self.acls.set_acls(broker_protocol, self.kafka, self.zk, self.topic, self.group)
         self.bounce()
diff --git a/tests/kafkatest/tests/zookeeper_security_upgrade_test.py b/tests/kafkatest/tests/zookeeper_security_upgrade_test.py
index 533ad366931..3bfc4781bf0 100644
--- a/tests/kafkatest/tests/zookeeper_security_upgrade_test.py
+++ b/tests/kafkatest/tests/zookeeper_security_upgrade_test.py
@@ -101,7 +101,7 @@ def test_zk_security_upgrade(self, security_protocol):
 
         # set acls
         if self.is_secure:
-            self.kafka.authorizer_class_name = "kafka.security.auth.SimpleAclAuthorizer"
+            self.kafka.authorizer_class_name = KafkaService.SIMPLE_AUTHORIZER
             self.acls.set_acls(security_protocol, self.kafka, self.zk, self.topic, self.group)
 
         if(self.no_sasl):
