diff --git a/client/src/main/java/com/metamx/druid/query/Queries.java b/client/src/main/java/com/metamx/druid/query/Queries.java
index fc33a0c2b74..ae5f958b563 100644
--- a/client/src/main/java/com/metamx/druid/query/Queries.java
+++ b/client/src/main/java/com/metamx/druid/query/Queries.java
@@ -64,7 +64,7 @@ public String apply(@Nullable AggregatorFactory input)
 
         Preconditions.checkArgument(
             missing.isEmpty(),
-            String.format("Missing fields [%s] for postAggregator [%s]", Joiner.on(",").join(missing), postAgg.getName())
+            "Missing fields [%s] for postAggregator [%s]", missing, postAgg.getName()
         );
         combinedAggNames.add(postAgg.getName());
       }
