diff --git a/example/quantization/launch_inference.sh b/example/quantization/launch_inference.sh
index 6ec8bc718944..8c839ba0f611 100755
--- a/example/quantization/launch_inference.sh
+++ b/example/quantization/launch_inference.sh
@@ -17,6 +17,7 @@
 # specific language governing permissions and limitations
 # under the License.
 
+set -ex
 
 python imagenet_inference.py --symbol-file=./model/imagenet1k-resnet-152-symbol.json --param-file=./model/imagenet1k-resnet-152-0000.params --rgb-mean=0,0,0 --num-skipped-batches=50 --num-inference-batches=500 --dataset=./data/val_256_q90.rec
 
diff --git a/example/quantization/launch_quantize.sh b/example/quantization/launch_quantize.sh
index edbf5457ab56..9aa4bee4bff1 100755
--- a/example/quantization/launch_quantize.sh
+++ b/example/quantization/launch_quantize.sh
@@ -17,6 +17,7 @@
 # specific language governing permissions and limitations
 # under the License.
 
+set -ex
 
 python imagenet_gen_qsym.py --model=imagenet1k-resnet-152 --calib-mode=none
 
diff --git a/src/operator/quantization/quantized_conv.cc b/src/operator/quantization/quantized_conv.cc
index c2b45703ec45..d7dc9fe4dbd8 100644
--- a/src/operator/quantization/quantized_conv.cc
+++ b/src/operator/quantization/quantized_conv.cc
@@ -28,8 +28,6 @@
 namespace mxnet {
 namespace op {
 
-// TODO(junwu): Reuse the InferShape function of convolution op after
-// this pr is merged: https://github.com/apache/incubator-mxnet/pull/8302
 bool QuantizedConvShape(const nnvm::NodeAttrs& attrs,
                         std::vector<TShape>* in_shape,
                         std::vector<TShape>* out_shape) {
