diff --git a/rust/Cargo.toml b/rust/Cargo.toml
index 826c6249852e..987696c6670f 100644
--- a/rust/Cargo.toml
+++ b/rust/Cargo.toml
@@ -16,7 +16,7 @@
 # under the License.
 
 [package]
-name = "apache-arrow"
+name = "arrow"
 version = "0.1.0"
 description = "Rust implementation of Apache Arrow"
 homepage = "https://github.com/apache/arrow"
@@ -34,4 +34,4 @@ name = "arrow"
 path = "src/lib.rs"
 
 [dependencies]
-bytes = "*"
\ No newline at end of file
+bytes = "*"
