diff --git a/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala b/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala
index 55f291236ef32..f22b5eaa90a5d 100644
--- a/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala
+++ b/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala
@@ -125,7 +125,7 @@ class DAGScheduler(
 
   /** Broadcast the serialized tasks only when they are bigger than it */
   private val broadcastTaskMinSize =
-    sc.getConf.getInt("spark.scheduler.broadcastTaskMinSize", 8) * 1024
+    sc.getConf.getInt("spark.scheduler.broadcastTaskMinSizeKB", 8) * 1024
 
   private def initializeEventProcessActor() {
     // blocking the thread until supervisor is started, which ensures eventProcessActor is
