diff --git a/superset-frontend/src/components/ErrorMessage/DatabaseErrorMessage.tsx b/superset-frontend/src/components/ErrorMessage/DatabaseErrorMessage.tsx
index c9ffb86e607c7..6968163aa7b59 100644
--- a/superset-frontend/src/components/ErrorMessage/DatabaseErrorMessage.tsx
+++ b/superset-frontend/src/components/ErrorMessage/DatabaseErrorMessage.tsx
@@ -29,19 +29,16 @@ interface DatabaseErrorExtra {
     code: number;
     message: string;
   }[];
-  engine_name: string;
+  engine_name: string | null;
 }
 
 function DatabaseErrorMessage({
   error,
-  source,
+  source = 'dashboard',
 }: ErrorMessageComponentProps<DatabaseErrorExtra>) {
   const { extra, level, message } = error;
 
-  const isVisualization = (['dashboard', 'explore'] as (
-    | string
-    | undefined
-  )[]).includes(source);
+  const isVisualization = ['dashboard', 'explore'].includes(source);
 
   const body = (
     <>
@@ -81,7 +78,7 @@ ${extra.issue_codes.map(issueCode => issueCode.message).join('\n')}`;
 
   return (
     <ErrorAlert
-      title={t('%s Error', extra.engine_name)}
+      title={t('%s Error', extra.engine_name || t('DB Engine'))}
       subtitle={message}
       level={level}
       source={source}
diff --git a/superset-frontend/src/components/ErrorMessage/ErrorAlert.tsx b/superset-frontend/src/components/ErrorMessage/ErrorAlert.tsx
index 10368a0c29548..8dd832e881e99 100644
--- a/superset-frontend/src/components/ErrorMessage/ErrorAlert.tsx
+++ b/superset-frontend/src/components/ErrorMessage/ErrorAlert.tsx
@@ -96,17 +96,14 @@ export default function ErrorAlert({
   body,
   copyText,
   level,
-  source,
+  source = 'dashboard',
   subtitle,
   title,
 }: ErrorAlertProps) {
   const [isModalOpen, setIsModalOpen] = useState(false);
   const [isBodyExpanded, setIsBodyExpanded] = useState(false);
 
-  const isExpandable = (['explore', 'sqllab'] as (
-    | string
-    | undefined
-  )[]).includes(source);
+  const isExpandable = ['explore', 'sqllab'].includes(source);
 
   return (
     <ErrorAlertDiv level={level}>
diff --git a/superset/db_engine_specs/athena.py b/superset/db_engine_specs/athena.py
index 81830de090629..91e1069ff68b2 100644
--- a/superset/db_engine_specs/athena.py
+++ b/superset/db_engine_specs/athena.py
@@ -23,7 +23,7 @@
 
 class AthenaEngineSpec(BaseEngineSpec):
     engine = "awsathena"
-    engine_name = "AWS Athena"
+    engine_name = "Amazon Athena"
 
     _time_grain_expressions = {
         None: "{col}",
diff --git a/superset/db_engine_specs/base.py b/superset/db_engine_specs/base.py
index 42489e2d85557..fdaee05cca986 100644
--- a/superset/db_engine_specs/base.py
+++ b/superset/db_engine_specs/base.py
@@ -137,7 +137,9 @@ class BaseEngineSpec:  # pylint: disable=too-many-public-methods
     """Abstract class for database engine specific configurations"""
 
     engine = "base"  # str as defined in sqlalchemy.engine.engine
-    engine_name = "Base Database"
+    engine_name: Optional[
+        str
+    ] = None  # used for user messages, overridden in child classes
     _date_trunc_functions: Dict[str, str] = {}
     _time_grain_expressions: Dict[Optional[str], str] = {}
     time_groupby_inline = False
diff --git a/superset/db_engine_specs/cockroachdb.py b/superset/db_engine_specs/cockroachdb.py
index bbb7604a1b604..f2f00c1a04784 100644
--- a/superset/db_engine_specs/cockroachdb.py
+++ b/superset/db_engine_specs/cockroachdb.py
@@ -19,4 +19,4 @@
 
 class CockroachDbEngineSpec(PostgresEngineSpec):
     engine = "cockroachdb"
-    engine_name = "Cockroach DB"
+    engine_name = "CockroachDB"
diff --git a/superset/db_engine_specs/druid.py b/superset/db_engine_specs/druid.py
index 13745740849f9..f941fd1bad7b0 100644
--- a/superset/db_engine_specs/druid.py
+++ b/superset/db_engine_specs/druid.py
@@ -35,7 +35,7 @@ class DruidEngineSpec(BaseEngineSpec):  # pylint: disable=abstract-method
     """Engine spec for Druid.io"""
 
     engine = "druid"
-    engine_name = "Druid"
+    engine_name = "Apache Druid"
     allows_joins = False
     allows_subqueries = True
 
diff --git a/superset/db_engine_specs/hana.py b/superset/db_engine_specs/hana.py
index 291a15946ce93..c4b157a624904 100644
--- a/superset/db_engine_specs/hana.py
+++ b/superset/db_engine_specs/hana.py
@@ -24,7 +24,7 @@
 
 class HanaEngineSpec(PostgresBaseEngineSpec):
     engine = "hana"
-    engine_name = "Hana"
+    engine_name = "SAP HANA"
     limit_method = LimitMethod.WRAP_SQL
     force_column_alias_quotes = True
     max_column_name_length = 30
diff --git a/superset/db_engine_specs/hive.py b/superset/db_engine_specs/hive.py
index 5e45cf640d6c7..baf94c69a049a 100644
--- a/superset/db_engine_specs/hive.py
+++ b/superset/db_engine_specs/hive.py
@@ -55,7 +55,7 @@ class HiveEngineSpec(PrestoEngineSpec):
     """Reuses PrestoEngineSpec functionality."""
 
     engine = "hive"
-    engine_name = "Hive"
+    engine_name = "Apache Hive"
     max_column_name_length = 767
     # pylint: disable=line-too-long
     _time_grain_expressions = {
diff --git a/superset/db_engine_specs/impala.py b/superset/db_engine_specs/impala.py
index c0e2e1c423165..9d8dc91374f4e 100644
--- a/superset/db_engine_specs/impala.py
+++ b/superset/db_engine_specs/impala.py
@@ -27,7 +27,7 @@ class ImpalaEngineSpec(BaseEngineSpec):
     """Engine spec for Cloudera's Impala"""
 
     engine = "impala"
-    engine_name = "Impala"
+    engine_name = "Apache Impala"
 
     _time_grain_expressions = {
         None: "{col}",
diff --git a/superset/db_engine_specs/kylin.py b/superset/db_engine_specs/kylin.py
index 34ad537b43d80..4a64091d155db 100644
--- a/superset/db_engine_specs/kylin.py
+++ b/superset/db_engine_specs/kylin.py
@@ -25,7 +25,7 @@ class KylinEngineSpec(BaseEngineSpec):  # pylint: disable=abstract-method
     """Dialect for Apache Kylin"""
 
     engine = "kylin"
-    engine_name = "Kylin"
+    engine_name = "Apache Kylin"
 
     _time_grain_expressions = {
         None: "{col}",
diff --git a/superset/db_engine_specs/pinot.py b/superset/db_engine_specs/pinot.py
index cb70a6174e8cf..d79a6947adefe 100644
--- a/superset/db_engine_specs/pinot.py
+++ b/superset/db_engine_specs/pinot.py
@@ -24,7 +24,7 @@
 
 class PinotEngineSpec(BaseEngineSpec):  # pylint: disable=abstract-method
     engine = "pinot"
-    engine_name = "Pinot"
+    engine_name = "Apache Pinot"
     allows_subqueries = False
     allows_joins = False
     allows_column_aliases = False
diff --git a/superset/db_engine_specs/postgres.py b/superset/db_engine_specs/postgres.py
index 72df7797aa7d9..0ccf51dba077f 100644
--- a/superset/db_engine_specs/postgres.py
+++ b/superset/db_engine_specs/postgres.py
@@ -38,7 +38,7 @@ class PostgresBaseEngineSpec(BaseEngineSpec):
     """ Abstract class for Postgres 'like' databases """
 
     engine = ""
-    engine_name = "Postgres"
+    engine_name = "PostgreSQL"
 
     _time_grain_expressions = {
         None: "{col}",
diff --git a/superset/db_engine_specs/redshift.py b/superset/db_engine_specs/redshift.py
index c2bc28c1fed7e..ef7b0e29a3eee 100644
--- a/superset/db_engine_specs/redshift.py
+++ b/superset/db_engine_specs/redshift.py
@@ -19,7 +19,7 @@
 
 class RedshiftEngineSpec(PostgresBaseEngineSpec):
     engine = "redshift"
-    engine_name = "Redshift"
+    engine_name = "Amazon Redshift"
     max_column_name_length = 127
 
     @staticmethod
diff --git a/superset/db_engine_specs/teradata.py b/superset/db_engine_specs/teradata.py
index 73788fc41dfa2..88bffa5a0752e 100644
--- a/superset/db_engine_specs/teradata.py
+++ b/superset/db_engine_specs/teradata.py
@@ -21,7 +21,7 @@ class TeradataEngineSpec(BaseEngineSpec):
     """Dialect for Teradata DB."""
 
     engine = "teradata"
-    engine_name = "Teradata DB"
+    engine_name = "Teradata"
     limit_method = LimitMethod.WRAP_SQL
     max_column_name_length = 30  # since 14.10 this is 128
 
