diff --git a/airflow/contrib/operators/dataflow_operator.py b/airflow/contrib/operators/dataflow_operator.py
index b7103eb43ddbb..6fd23f18fbc09 100644
--- a/airflow/contrib/operators/dataflow_operator.py
+++ b/airflow/contrib/operators/dataflow_operator.py
@@ -155,8 +155,8 @@ def __init__(
 
         https://cloud.google.com/dataflow/pipelines/specifying-exec-params
 
-        :param py_file: Reference to the python dataflow pipleline file, e.g.,
-            /some/local/file/path/to/your/python/pipeline/file.py.
+        :param py_file: Reference to the python dataflow pipleline file.py, e.g.,
+            /some/local/file/path/to/your/python/pipeline/file.
         :type py_file: string
         :param py_options: Additional python options.
         :type pyt_options: list of strings, e.g., ["-m", "-v"].
