diff --git a/src/relay/pass/type_infer.cc b/src/relay/pass/type_infer.cc
index c1d1a66d3164..bc84bddaad79 100644
--- a/src/relay/pass/type_infer.cc
+++ b/src/relay/pass/type_infer.cc
@@ -315,7 +315,7 @@ class TypeInferencer : private ExprFunctor<Type(const Expr&)>,
         ss << "match expression does not handle the following cases: ";
         int i = 0;
         for (auto cs : unmatched_cases) {
-          ss << "case " << i << ": \n" << PrettyPrint(cs);
+          ss << "case " << i++ << ": \n" << PrettyPrint(cs);
         }
         this->ReportFatalError(
           match,
