diff --git a/hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java b/hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java
index 93a3c83c0b0..b778e0397cb 100644
--- a/hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java
+++ b/hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java
@@ -18,7 +18,7 @@
 
 package org.apache.hadoop.hdds.client;
 
-import org.apache.hadoop.ozone.OzoneConsts;
+import com.google.common.base.Strings;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -117,14 +117,6 @@ public Units getUnit() {
     return this.rawQuotaInBytes.getUnit();
   }
 
-  /**
-   * Constructs a default Quota object.
-   */
-  private OzoneQuota() {
-    this.quotaInCounts = OzoneConsts.QUOTA_RESET;
-    this.quotaInBytes = OzoneConsts.QUOTA_RESET;
-  }
-
   /**
    * Constructor for Ozone Quota.
    *
@@ -144,7 +136,7 @@ private OzoneQuota(long quotaInCounts, RawQuotaInBytes rawQuotaInBytes) {
    * @return string representation of quota
    */
   public static String formatQuota(OzoneQuota quota) {
-    return String.valueOf(quota.getRawSize())+ quota.getUnit();
+    return String.valueOf(quota.getRawSize()) + quota.getUnit();
   }
 
   /**
@@ -159,7 +151,7 @@ public static String formatQuota(OzoneQuota quota) {
   public static OzoneQuota parseQuota(String quotaInBytes,
       long quotaInCounts) {
 
-    if ((quotaInBytes == null) || (quotaInBytes.isEmpty())) {
+    if (Strings.isNullOrEmpty(quotaInBytes)) {
       throw new IllegalArgumentException(
           "Quota string cannot be null or empty.");
     }
@@ -267,7 +259,7 @@ public long getQuotaInBytes() {
 
   @Override
   public String toString() {
-    return "Bytes Quota: " + rawQuotaInBytes.toString() + "\n" +
+    return "Space Bytes Quota: " + rawQuotaInBytes.toString() + "\n" +
         "Counts Quota: " + quotaInCounts;
   }
 }
diff --git a/hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConsts.java b/hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConsts.java
index 1dec4d797ff..6f8de0f205f 100644
--- a/hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConsts.java
+++ b/hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConsts.java
@@ -205,7 +205,7 @@ public static Versioning getVersioning(boolean versioning) {
   public static final long MAX_QUOTA_IN_BYTES = 1024L * 1024 * TB;
 
   /**
-   * Quota RESET default is -1, which means it does not take effect.
+   * Quota RESET default is -1, which means quota is not set.
    */
   public static final long QUOTA_RESET = -1;
 
