diff --git a/fe/fe-core/src/main/java/org/apache/doris/http/meta/MetaService.java b/fe/fe-core/src/main/java/org/apache/doris/http/meta/MetaService.java
index f3bf61036909..964cbf1d1e7d 100644
--- a/fe/fe-core/src/main/java/org/apache/doris/http/meta/MetaService.java
+++ b/fe/fe-core/src/main/java/org/apache/doris/http/meta/MetaService.java
@@ -180,7 +180,7 @@ public void executeGet(BaseRequest request, BaseResponse response) {
             if (Catalog.getCurrentCatalog().isMaster()) {
                 response.appendContent("this node is master, reject image put");
                 writeResponse(request, response, HttpResponseStatus.BAD_REQUEST);
-                LOG.error("this node is master, but receive image put from host{}, reject it", machine);
+                LOG.error("this node is master, but receive image put from host {}, reject it", machine);
                 return;
             }
 
@@ -188,7 +188,8 @@ public void executeGet(BaseRequest request, BaseResponse response) {
             // if accepted, newly added log will not be replayed when restart
             long maxJournalId = Catalog.getCurrentCatalog().getMaxJournalId();
             if (version > maxJournalId) {
-                response.appendContent("image version is bigger than local max journal id, reject image put");
+                response.appendContent("image version [" + version + "] is bigger than local max journal id ["
+                                               + maxJournalId + "], reject image put");
                 writeResponse(request, response, HttpResponseStatus.BAD_REQUEST);
                 LOG.error("receive image whose version [{}] is bigger than local max journal id [{}], reject it",
                           version, maxJournalId);
