diff --git a/apm-sniffer/apm-agent-core/src/test/java/org/apache/skywalking/apm/agent/core/context/CorrelationContextTest.java b/apm-sniffer/apm-agent-core/src/test/java/org/apache/skywalking/apm/agent/core/context/CorrelationContextTest.java
index 29979b6225b..32e5b5f3556 100644
--- a/apm-sniffer/apm-agent-core/src/test/java/org/apache/skywalking/apm/agent/core/context/CorrelationContextTest.java
+++ b/apm-sniffer/apm-agent-core/src/test/java/org/apache/skywalking/apm/agent/core/context/CorrelationContextTest.java
@@ -23,7 +23,7 @@
 import org.apache.skywalking.apm.agent.core.boot.ServiceManager;
 import org.apache.skywalking.apm.agent.core.conf.Config;
 import org.apache.skywalking.apm.agent.core.context.tag.StringTag;
-import org.apache.skywalking.apm.agent.core.context.trace.AbstractSpan;
+import org.apache.skywalking.apm.agent.core.context.trace.AbstractTracingSpan;
 import org.apache.skywalking.apm.agent.core.context.trace.TraceSegment;
 import org.apache.skywalking.apm.agent.core.context.util.TagValuePair;
 import org.apache.skywalking.apm.agent.core.test.tools.AgentServiceRule;
@@ -150,24 +150,23 @@ public void testDeserialize() {
     }
 
     @Test
-    public void testHandleWhenAutoTagKeysEmpty() {
+    public void testHandleWhenAutoTagKeysNotEmpty() {
         ContextManager.createEntrySpan("/testFirstEntry", new ContextCarrier());
-        ContextManager.getCorrelationContext().put("a", "b");
+        ContextManager.getCorrelationContext().put("autotag", "b");
         ContextManager.stopSpan();
         TraceSegment traceSegment = tracingData.getTraceSegments().get(0);
-        List<AbstractSpan> spans = Whitebox.getInternalState(traceSegment, "spans");
-        Assert.assertNull(Whitebox.getInternalState(spans.get(0), "tags"));
+        List<AbstractTracingSpan> spans = Whitebox.getInternalState(traceSegment, "spans");
+        List<TagValuePair> tags = Whitebox.getInternalState(spans.get(0), "tags");
+        Assert.assertEquals(new TagValuePair(new StringTag("autotag"), "b"), tags.get(0));
     }
 
     @Test
-    public void testHandleWhenAutoTagKeysNotEmpty() {
+    public void testHandleWhenAutoTagKeysEmpty() {
         ContextManager.createEntrySpan("/testFirstEntry", new ContextCarrier());
-        ContextManager.getCorrelationContext().put("autotag", "b");
+        ContextManager.getCorrelationContext().put("a", "b");
         ContextManager.stopSpan();
         TraceSegment traceSegment = tracingData.getTraceSegments().get(0);
-        List<AbstractSpan> spans = Whitebox.getInternalState(traceSegment, "spans");
-        List<TagValuePair> tags = Whitebox.getInternalState(spans.get(0), "tags");
-        Assert.assertEquals(1, tags.size());
-        Assert.assertEquals(new TagValuePair(new StringTag("autotag"), "b"), tags.get(0));
+        List<AbstractTracingSpan> spans = Whitebox.getInternalState(traceSegment, "spans");
+        Assert.assertNull(Whitebox.getInternalState(spans.get(0), "tags"));
     }
 }
