diff --git a/python/mxnet/test_utils.py b/python/mxnet/test_utils.py
index 0bb28a0ef13a..e98bbdef7c78 100644
--- a/python/mxnet/test_utils.py
+++ b/python/mxnet/test_utils.py
@@ -1379,7 +1379,7 @@ def list_gpus():
     for cmd in nvidia_smi:
         try:
             re = subprocess.check_output([cmd, "-L"], universal_newlines=True)
-        except OSError:
+        except (subprocess.CalledProcessError, OSError) as e:
             pass
     return range(len([i for i in re.split('\n') if 'GPU' in i]))
 
