diff --git a/daffodil-core/src/main/scala/org/apache/daffodil/dsom/RestrictionUnion.scala b/daffodil-core/src/main/scala/org/apache/daffodil/dsom/RestrictionUnion.scala
index 35ce2a252e..68317f7bfc 100644
--- a/daffodil-core/src/main/scala/org/apache/daffodil/dsom/RestrictionUnion.scala
+++ b/daffodil-core/src/main/scala/org/apache/daffodil/dsom/RestrictionUnion.scala
@@ -76,7 +76,8 @@ final class Restriction(xmlArg: Node, val simpleTypeDef: SimpleTypeDefBase)
     val baseQNameString = baseQNameNodeSeq.text
     val tryBaseQName = QName.resolveRef(baseQNameString, xml.scope,
       tunable.unqualifiedPathStepPolicy)
-    Assert.invariant(tryBaseQName.isSuccess)
+    schemaDefinitionUnless(tryBaseQName.isSuccess,
+      "Failed to resolve base property reference for xs:restriction: " + tryBaseQName.failed.get.getMessage)
     tryBaseQName.get
   }
 
diff --git a/daffodil-test/src/test/resources/org/apache/daffodil/section05/simple_types/SimpleTypes.tdml b/daffodil-test/src/test/resources/org/apache/daffodil/section05/simple_types/SimpleTypes.tdml
index a9e6a35661..19d4ae66a8 100644
--- a/daffodil-test/src/test/resources/org/apache/daffodil/section05/simple_types/SimpleTypes.tdml
+++ b/daffodil-test/src/test/resources/org/apache/daffodil/section05/simple_types/SimpleTypes.tdml
@@ -4354,6 +4354,24 @@
     </tdml:errors>
   </tdml:unparserTestCase>
 
+  <tdml:defineSchema name="SimpleTypes-restriction-base" elementFormDefault="unqualified">
+    <xs:include schemaLocation="org/apache/daffodil/xsd/DFDLGeneralFormat.dfdl.xsd"/>
+    <dfdl:format ref="ex:GeneralFormat"/>
+
+    <xs:element name="embeddedLeadingSpace">
+      <xs:simpleType>
+        <xs:restriction base=" xs:string" />
+      </xs:simpleType>
+    </xs:element>
+
+    <xs:element name="embeddedTrailingSpace">
+      <xs:simpleType>
+        <xs:restriction base="xs:string " />
+      </xs:simpleType>
+    </xs:element>
+
+  </tdml:defineSchema>
+
   <tdml:defineSchema name="SimpleTypes-padding">
     <xs:include schemaLocation="org/apache/daffodil/xsd/DFDLGeneralFormat.dfdl.xsd"/>
     <dfdl:format ref="ex:GeneralFormat" representation="text"
@@ -4468,7 +4486,39 @@
     <xs:element name="string12" dfdl:textStringPadCharacter="%" dfdl:lengthKind="delimited" type="xs:string" />
 
   </tdml:defineSchema>
-  
+
+  <!--
+  Test name: restrictionBaseEmbeddedTrailingSpace
+  Schema: SimpleTypes-restriction-base
+  Purpose: This document demonstrates the SDE due to trailing space in restriction base
+  -->
+
+  <tdml:parserTestCase name="restrictionBaseEmbeddedTrailingSpace" root="embeddedTrailingSpace"
+                       model="SimpleTypes-restriction-base"
+                       description="Test SDE due to trailing space in base">
+    <tdml:document>test</tdml:document>
+    <tdml:errors>
+      <tdml:error>Schema Definition Error</tdml:error>
+      <tdml:error>Failed to resolve base property reference for xs:restriction:</tdml:error>
+    </tdml:errors>
+  </tdml:parserTestCase>
+
+  <!--
+  Test name: restrictionBaseEmbeddedLeadingSpace
+  Schema: SimpleTypes-restriction-base
+  Purpose: This document demonstrates the SDE due to leading space in restriction base
+  -->
+
+  <tdml:parserTestCase name="restrictionBaseEmbeddedLeadingSpace" root="embeddedLeadingSpace"
+                       model="SimpleTypes-restriction-base"
+                       description="Test SDE due to leading space in base">
+    <tdml:document>test</tdml:document>
+    <tdml:errors>
+      <tdml:error>Schema Definition Error</tdml:error>
+      <tdml:error>Failed to resolve base property reference for xs:restriction:</tdml:error>
+    </tdml:errors>
+  </tdml:parserTestCase>
+
   <!--
     Test name: literalChar_padding
     Schema: SimpleTypes-padding
diff --git a/daffodil-test/src/test/scala/org/apache/daffodil/section05/simple_types/TestSimpleTypes2.scala b/daffodil-test/src/test/scala/org/apache/daffodil/section05/simple_types/TestSimpleTypes2.scala
index 58027877ad..16715b0be5 100644
--- a/daffodil-test/src/test/scala/org/apache/daffodil/section05/simple_types/TestSimpleTypes2.scala
+++ b/daffodil-test/src/test/scala/org/apache/daffodil/section05/simple_types/TestSimpleTypes2.scala
@@ -99,4 +99,7 @@ class TestSimpleTypes2 {
 
   @Test def test_hexBinary_Delimited_04(): Unit = { runner.runOneTest("hexBinary_Delimited_04") }
 
+  @Test def test_restrictionBaseEmbeddedLeadlingSpace(): Unit = { runner.runOneTest("restrictionBaseEmbeddedLeadingSpace") }
+  @Test def test_restrictionBaseEmbeddedTrailingSpace(): Unit = { runner.runOneTest("restrictionBaseEmbeddedTrailingSpace") }
+
 }
