diff --git a/api/test/e2e/base.go b/api/test/e2e/base.go
index 8f4116acb3..99c763d259 100644
--- a/api/test/e2e/base.go
+++ b/api/test/e2e/base.go
@@ -31,6 +31,7 @@ import (
 )
 
 var token string
+var APISIXInternalUrl = "http://172.16.238.30:9080"
 
 func init() {
 	//login to get auth token
diff --git a/api/test/e2e/route_online_debug_test.go b/api/test/e2e/route_online_debug_test.go
index be41027b85..4dd88f2a50 100644
--- a/api/test/e2e/route_online_debug_test.go
+++ b/api/test/e2e/route_online_debug_test.go
@@ -42,7 +42,7 @@ func TestRoute_Online_Debug_Route_Not_Exist(t *testing.T) {
 		testCaseCheck(tc)
 	}
 	basepath := "http://127.0.0.1:9000/apisix/admin/debug-request-forwarding"
-	request, _ := http.NewRequest("POST", basepath, strings.NewReader(`{"url": "http://172.16.238.30:9080/hello_","method": "GET","request_protocol": "http"}`))
+	request, _ := http.NewRequest("POST", basepath, strings.NewReader(`{"url": "`+APISIXInternalUrl+`/hello_","method": "GET","request_protocol": "http"}`))
 	request.Header.Add("Authorization", token)
 	resp, err := http.DefaultClient.Do(request)
 	if err != nil {
@@ -94,7 +94,7 @@ func TestRoute_Online_Debug_Route_With_Query_Params(t *testing.T) {
 			Method:   http.MethodPost,
 			Path:     "/apisix/admin/debug-request-forwarding",
 			Body: `{
-				"url": "http://172.16.238.30:9080/hello?name=aaa",
+				"url": "` + APISIXInternalUrl + `/hello?name=aaa",
 				"request_protocol": "http",
 				"method": "GET"
 			}`,
@@ -166,7 +166,7 @@ func TestRoute_Online_Debug_Route_With_Header_Params(t *testing.T) {
 			Method:   http.MethodPost,
 			Path:     "/apisix/admin/debug-request-forwarding",
 			Body: `{
-				"url": "http://172.16.238.30:9080/hello",
+				"url": "` + APISIXInternalUrl + `/hello",
 				"request_protocol": "http",
 				"method": "GET",
 				"header_params": {
@@ -238,7 +238,7 @@ func TestRoute_Online_Debug_Route_With_Body_Params(t *testing.T) {
 			Method:   http.MethodPost,
 			Path:     "/apisix/admin/debug-request-forwarding",
 			Body: `{
-				"url": "http://172.16.238.30:9080/hello",
+				"url": "` + APISIXInternalUrl + `/hello",
 				"request_protocol": "http",
 				"method": "POST",
 				"body_params": {
@@ -342,7 +342,7 @@ func TestRoute_Online_Debug_Route_With_Basic_Auth(t *testing.T) {
 			Method:   http.MethodPost,
 			Path:     "/apisix/admin/debug-request-forwarding",
 			Body: `{
-				"url": "http://172.16.238.30:9080/hello",
+				"url": "` + APISIXInternalUrl + `/hello",
 				"request_protocol": "http",
 				"method": "GET",
 				"header_params": {
@@ -360,7 +360,7 @@ func TestRoute_Online_Debug_Route_With_Basic_Auth(t *testing.T) {
 
 	// online debug without basic-auth
 	basepath := "http://127.0.0.1:9000/apisix/admin/debug-request-forwarding"
-	request, _ := http.NewRequest("POST", basepath, strings.NewReader(`{"url": "http://172.16.238.30:9080/hello","method": "GET","request_protocol":"http"}`))
+	request, _ := http.NewRequest("POST", basepath, strings.NewReader(`{"url": "`+APISIXInternalUrl+`/hello","method": "GET","request_protocol":"http"}`))
 	request.Header.Add("Authorization", token)
 	resp, err := http.DefaultClient.Do(request)
 	if err != nil {
@@ -500,7 +500,7 @@ func TestRoute_Online_Debug_Route_With_Jwt_Auth(t *testing.T) {
 			Method:   http.MethodPost,
 			Path:     "/apisix/admin/debug-request-forwarding",
 			Body: `{
-				"url": "http://172.16.238.30:9080/hello",
+				"url": "` + APISIXInternalUrl + `/hello",
 				"request_protocol": "http",
 				"method": "GET",
 				"header_params": {
@@ -518,7 +518,7 @@ func TestRoute_Online_Debug_Route_With_Jwt_Auth(t *testing.T) {
 
 	// online debug without jwt-auth
 	basepath := "http://127.0.0.1:9000/apisix/admin/debug-request-forwarding"
-	request, _ := http.NewRequest("POST", basepath, strings.NewReader(`{"url": "http://172.16.238.30:9080/hello","method": "GET","request_protocol":"http"}`))
+	request, _ := http.NewRequest("POST", basepath, strings.NewReader(`{"url": "`+APISIXInternalUrl+`/hello","method": "GET","request_protocol":"http"}`))
 	request.Header.Add("Authorization", token)
 	resp, err := http.DefaultClient.Do(request)
 	if err != nil {
@@ -637,7 +637,7 @@ func TestRoute_Online_Debug_Route_With_Key_Auth(t *testing.T) {
 			Method:   http.MethodPost,
 			Path:     "/apisix/admin/debug-request-forwarding",
 			Body: `{
-				"url": "http://172.16.238.30:9080/hello",
+				"url": "` + APISIXInternalUrl + `/hello",
 				"request_protocol": "http",
 				"method": "GET",
 				"header_params": {
@@ -655,7 +655,7 @@ func TestRoute_Online_Debug_Route_With_Key_Auth(t *testing.T) {
 
 	// online debug without key-auth
 	basepath := "http://127.0.0.1:9000/apisix/admin/debug-request-forwarding"
-	request, _ := http.NewRequest("POST", basepath, strings.NewReader(`{"url": "http://172.16.238.30:9080/hello","method": "GET","request_protocol": "http"}`))
+	request, _ := http.NewRequest("POST", basepath, strings.NewReader(`{"url": "`+APISIXInternalUrl+`/hello","method": "GET","request_protocol": "http"}`))
 	request.Header.Add("Authorization", token)
 	resp, err := http.DefaultClient.Do(request)
 	if err != nil {
@@ -777,7 +777,7 @@ func TestRoute_Online_Debug_Route_With_Query_Params_Key_Auth(t *testing.T) {
 			Method:   http.MethodPost,
 			Path:     "/apisix/admin/debug-request-forwarding",
 			Body: `{
-				"url": "http://172.16.238.30:9080/hello?name=aaa",
+				"url": "` + APISIXInternalUrl + `/hello?name=aaa",
 				"request_protocol": "http",
 				"method": "GET",
 				"header_params": {
@@ -795,7 +795,7 @@ func TestRoute_Online_Debug_Route_With_Query_Params_Key_Auth(t *testing.T) {
 
 	// online debug without key-auth
 	basepath := "http://127.0.0.1:9000/apisix/admin/debug-request-forwarding"
-	request, _ := http.NewRequest("POST", basepath, strings.NewReader(`{"url": "http://172.16.238.30:9080/hello?name=aaa","method": "GET","request_protocol": "http"}`))
+	request, _ := http.NewRequest("POST", basepath, strings.NewReader(`{"url": "`+APISIXInternalUrl+`/hello?name=aaa","method": "GET","request_protocol": "http"}`))
 	request.Header.Add("Authorization", token)
 	resp, err := http.DefaultClient.Do(request)
 	if err != nil {
