diff --git a/cppcache/integration/test/CMakeLists.txt b/cppcache/integration/test/CMakeLists.txt
index 034f694bbd..db198a4592 100644
--- a/cppcache/integration/test/CMakeLists.txt
+++ b/cppcache/integration/test/CMakeLists.txt
@@ -27,7 +27,7 @@ add_executable(cpp-integration-test
   ExampleTest.cpp
   ExpirationTest.cpp
   FunctionExecutionTest.cpp
-  PartitionRegionOpsWithRedundancyAndServerGoesDown.cpp
+  PartitionRegionOpsTest.cpp
   PdxInstanceTest.cpp
   PdxJsonTypeTest.cpp
   PdxSerializerTest.cpp
diff --git a/cppcache/integration/test/PartitionRegionOpsWithRedundancyAndServerGoesDown.cpp b/cppcache/integration/test/PartitionRegionOpsTest.cpp
similarity index 95%
rename from cppcache/integration/test/PartitionRegionOpsWithRedundancyAndServerGoesDown.cpp
rename to cppcache/integration/test/PartitionRegionOpsTest.cpp
index ad8eeb6412..51ea808b01 100644
--- a/cppcache/integration/test/PartitionRegionOpsWithRedundancyAndServerGoesDown.cpp
+++ b/cppcache/integration/test/PartitionRegionOpsTest.cpp
@@ -99,8 +99,7 @@ void getEntries(std::shared_ptr<Region> region, int numEntries) {
  * client metadata.
  */
 
-TEST(PartitionRegionOpsWithRedundancyAndServerGoesDown,
-     getPartitionedRegionWithRedundancyServerGoesDown) {
+TEST(PartitionRegionOpsTest, getPartitionedRegionWithRedundancyServerGoesDown) {
   Cluster cluster{LocatorCount{1}, ServerCount{2}};
   cluster.start();
   cluster.getGfsh()
@@ -139,8 +138,7 @@ TEST(PartitionRegionOpsWithRedundancyAndServerGoesDown,
  * When the server is brought back again, the meta data is refreshed
  * after putting again values.
  */
-TEST(PartitionRegionOpsWithRedundancyAndServerGoesDown,
-     putPartitionedRegionWithRedundancyServerGoesDown) {
+TEST(PartitionRegionOpsTest, putPartitionedRegionWithRedundancyServerGoesDown) {
   Cluster cluster{LocatorCount{1}, ServerCount{2}};
   cluster.start();
   cluster.getGfsh()
