diff --git a/heron/common/src/python/topology.py b/heron/common/src/python/topology.py
index fb2a5fbaa87..d3690381c3b 100644
--- a/heron/common/src/python/topology.py
+++ b/heron/common/src/python/topology.py
@@ -28,7 +28,6 @@ def __init__(self, name, state_manager_name):
     self.physical_plan = None
     self.execution_state = None
     self.id = None
-    self.dc = None
     self.cluster = None
     self.environ = None
     self.tmaster = None
@@ -110,7 +109,7 @@ def set_physical_plan(self, physical_plan):
   def get_execution_state_dc_environ(self, execution_state):
     """
     Helper function to extract dc and environ from execution_state.
-    Returns a tuple (dc, environ).
+    Returns a tuple (cluster, environ).
     """
     # TODO: This should be removed when old version of execution_state is no
     # longer in use.
