diff --git a/core/src/main/scala/org/apache/spark/api/r/RBackendHandler.scala b/core/src/main/scala/org/apache/spark/api/r/RBackendHandler.scala
index 0c1ad63142be3..0095548c463cc 100644
--- a/core/src/main/scala/org/apache/spark/api/r/RBackendHandler.scala
+++ b/core/src/main/scala/org/apache/spark/api/r/RBackendHandler.scala
@@ -224,7 +224,7 @@ private[r] class RBackendHandler(server: RBackend)
                 case _ => parameterType
               }
             }
-            if (!(!parameterType.isPrimitive && args(i) == null) &&
+            if ((parameterType.isPrimitive || args(i) != null) &&
                 !parameterWrapperType.isInstance(args(i))) {
               argMatched = false
             }
