diff --git a/plugins/metrics/src/main/java/org/apache/cloudstack/metrics/MetricsServiceImpl.java b/plugins/metrics/src/main/java/org/apache/cloudstack/metrics/MetricsServiceImpl.java
index 8b76a1770a3c..37f1f559483f 100644
--- a/plugins/metrics/src/main/java/org/apache/cloudstack/metrics/MetricsServiceImpl.java
+++ b/plugins/metrics/src/main/java/org/apache/cloudstack/metrics/MetricsServiceImpl.java
@@ -25,6 +25,7 @@
 import com.cloud.capacity.CapacityManager;
 import com.cloud.capacity.dao.CapacityDao;
 import com.cloud.capacity.dao.CapacityDaoImpl;
+import com.cloud.cluster.dao.ManagementServerHostDao;
 import com.cloud.dc.DataCenter;
 import com.cloud.dc.dao.ClusterDao;
 import com.cloud.dc.dao.DataCenterDao;
@@ -95,6 +96,8 @@ public class MetricsServiceImpl extends ComponentLifecycleBase implements Metric
     private DomainRouterDao domainRouterDao;
     @Inject
     private CapacityDao capacityDao;
+    @Inject
+    private ManagementServerHostDao managementServerHostDao;
 
     protected MetricsServiceImpl() {
         super();
@@ -138,6 +141,7 @@ public InfrastructureResponse listInfrastructure() {
             }
         }
         response.setCpuSockets(cpuSockets);
+        response.setManagementServers(managementServerHostDao.listAll().size());
         return response;
     }
 
diff --git a/plugins/metrics/src/main/java/org/apache/cloudstack/response/InfrastructureResponse.java b/plugins/metrics/src/main/java/org/apache/cloudstack/response/InfrastructureResponse.java
index a4db345d21d8..03b54abd2ec3 100644
--- a/plugins/metrics/src/main/java/org/apache/cloudstack/response/InfrastructureResponse.java
+++ b/plugins/metrics/src/main/java/org/apache/cloudstack/response/InfrastructureResponse.java
@@ -59,6 +59,10 @@ public class InfrastructureResponse extends BaseResponse {
     @Param(description = "Number of cpu sockets")
     private Integer cpuSockets;
 
+    @SerializedName("managementservers")
+    @Param(description = "Number of management servers")
+    private Integer managementServers;
+
     public InfrastructureResponse() {
         setObjectName("infrastructure");
     }
@@ -98,4 +102,8 @@ public void setRouters(final Integer routers) {
     public void setCpuSockets(final Integer cpuSockets) {
         this.cpuSockets = cpuSockets;
     }
+
+    public void setManagementServers(Integer managementServers) {
+        this.managementServers = managementServers;
+    }
 }
