diff --git a/CHANGELOG.asciidoc b/CHANGELOG.asciidoc
index eee27e95aaf..12e977c6568 100644
--- a/CHANGELOG.asciidoc
+++ b/CHANGELOG.asciidoc
@@ -22,8 +22,11 @@ TinkerPop 3.1.0 (A 187 On The Undercover Gremlinz)
 
 image::https://raw.githubusercontent.com/apache/incubator-tinkerpop/master/docs/static/images/gremlin-gangster.png[width=185]
 
-TinkerPop 3.1.0 (NOT OFFICIALLY RELEASED YET)
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+[[release-3.1.0-incubating]]
+TinkerPop 3.1.0 (Release Date: November 16, 2015)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This release also includes changes from <<release-3-0-1-incubating, 3.0.1-incubating>> and <<release-3-0-2-incubating, 3.0.2-incubating>>.
 
 * Fixed bug in Gryo and GraphSON (with embedded types) serialization for serialization of results returned from `Map.entrySet()`.
 * `Transaction` settings for `onReadWrite` and `onClose` are now `ThreadLocal` in nature of standard transactions.
@@ -90,11 +93,94 @@ TinkerPop 3.1.0 (NOT OFFICIALLY RELEASED YET)
 * Made `Path` iterable, so that it can be `unfold()`'ed and used by local steps like `min(local)`, `max(local)`, etc.
 * `WhereTraversalStep` and `WherePredicateStep` are now the only "special" `Scoping` steps after `MatchStartStep` in `match()`.
 
+Bugs
+^^^^
+
+* TINKERPOP3-774 order / dedup issues
+* TINKERPOP3-799 [Proposal] with()-modulator for stream level variable binding.
+* TINKERPOP3-801 groupCount() fails for vertices (elements?) (using Spark)
+* TINKERPOP3-811 AddPropertyStepTest fails "all of a sudden"
+* TINKERPOP3-823 addV() broken for multi-value properties
+* TINKERPOP3-843 Misspecified HADOOP_GREMLIN_LIBS generates NullPointerException
+* TINKERPOP3-857 Add GraphComputer.config(key,value)
+* TINKERPOP3-895 Use "as BinaryOperator" and remove GBinaryOperator
+* TINKERPOP3-903 Fix empty buffer return upon buffer capacity exceeded
+* TINKERPOP3-910 In session transaction opened from sessionless request
+* TINKERPOP3-918 ComputerVerificationStrategy is too restrictive
+* TINKERPOP3-926 Renamed TinkerGraph public statics to common pattern used for other statics.
+* TINKERPOP3-948 AbstractGremlinProcessTest.checkMap not asserted in GroupTest
+* TINKERPOP3-953 Artifact equality is not evaluating properly
+* TINKERPOP3-955 HashMap$Node not serializable
+
+Improvements
+^^^^^^^^^^^^
+
+* TINKERPOP3-297 Ensure Consistent Behavior Over Deleted Elements *(breaking)*
+* TINKERPOP3-333 Support VertexProperty in PartitionStrategy
+* TINKERPOP3-391 More fluency in GraphComputer for parameterization.
+* TINKERPOP3-616 Use Spark 1.3.0 in Hadoop-Gremlin.
+* TINKERPOP3-624 Passing Detached/Referenced to Graph.vertices/edge()
+* TINKERPOP3-680 Configurable Channelizer for Gremlin Driver
+* TINKERPOP3-728 Improve Remote Graph Object Treatment in Console
+* TINKERPOP3-756 Provide a strict parsing option for GraphMLReader
+* TINKERPOP3-760 Make loop counter accessible within repeat()
+* TINKERPOP3-762 Allow mid-traversal V() (and E())
+* TINKERPOP3-765 Decompose AbstractTransaction for different transactional contexts *(breaking)*
+* TINKERPOP3-767 Path should play well with "local" steps.
+* TINKERPOP3-768 MatchStep in OLAP should be smart about current vertex.
+* TINKERPOP3-769 Make the introduction of the TP3 docs story better.
+* TINKERPOP3-772 TraverserRequirement.LABELED_PATH
+* TINKERPOP3-796 Support merge binary operator for Gremlin sacks *(breaking)*
+* TINKERPOP3-798 [Proposal] Rename mapKeys()/mapValues() to select(keys) and select(values).
+* TINKERPOP3-802 Provide sack(object) so that the sack can be directly set.
+* TINKERPOP3-803 A better solution to g.V(someCollection.toArray())
+* TINKERPOP3-805 Enforce AutoCloseable Semantics on Transaction *(breaking)*
+* TINKERPOP3-821 Improve testing around TraversalHelper around recursive methods
+* TINKERPOP3-825 [Proposal] SetBulkStep (sideEffectStep)
+* TINKERPOP3-826 OneToManyBarrierStrategy
+* TINKERPOP3-827 Add a console session to the PageRank section of the docs.
+* TINKERPOP3-829 TinkerGraphComputer should support the user specified thread/worker count.
+* TINKERPOP3-835 Shade Jackson Dependencies *(breaking)*
+* TINKERPOP3-836 Support Hadoop2 in place of Hadoop1
+* TINKERPOP3-850 Reduce Graph.addVertex overload ambiguity *(breaking)*
+* TINKERPOP3-851 GroupCountStep needs a by() for the count.
+* TINKERPOP3-861 Solve "The Number Problem" for Operator (and follow on operators)
+* TINKERPOP3-863 [Proposal] Turn off bulking -- or is there something more general? (hope not).
+* TINKERPOP3-866 GroupStep and Traversal-Based Reductions *(breaking)*
+* TINKERPOP3-868 Allow Spark Gremlin Computer to Reuse Spark Contexts
+* TINKERPOP3-874 Rename Gremlin-Spark properties using gremlin.spark prefix. *(breaking)*
+* TINKERPOP3-876 Rename VendorOptimizationStrategy XXXOptimizationStrategy *(breaking)*
+* TINKERPOP3-879 Remove deprecated promoteBindings from GremlinExecutor *(breaking)*
+* TINKERPOP3-885 Change Transaction.onReadWrite() to be a ThreadLocal setting *(breaking)*
+* TINKERPOP3-888 GraphTraversal.property overloads *(breaking)*
+* TINKERPOP3-896 Simplify the {{withSack}} methods of {{GraphTraversalSource}}. *(breaking)*
+* TINKERPOP3-897 Remove deprecated GSupplier, GFunction, GConsumer, etc. methods. *(breaking)*
+* TINKERPOP3-898 Rename ConjuctionP and ConjuctionStep to ConnectiveP and ConnectiveStep *(breaking)*
+* TINKERPOP3-899 Bump to the latest version of Neo4j.
+* TINKERPOP3-900 Provide by(object) which compiles to by(constant(object))
+* TINKERPOP3-901 Option for use of Netty epoll on Linux to reduce GC pressure
+* TINKERPOP3-904 BulkLoaderVertexProgram optimizations
+* TINKERPOP3-905 Harden time oriented tests in ResultQueueTest
+* TINKERPOP3-907 getters for RepeatStep.untilTraversal and RepeatStep.emitTraversal
+* TINKERPOP3-908 Use line breaks in documentation
+* TINKERPOP3-909 Improve steps that handle numeric data
+* TINKERPOP3-911 Allow setting Thread Specific Spark JobGroup/Custom Properties based on hadoop conf
+* TINKERPOP3-913 Rename Gremlin Server arguments rebinding to alias
+* TINKERPOP3-914 DriverRemoteAcceptor in Gremlin Console supports aliases
+* TINKERPOP3-917 Add HadoopGraph.open(String)
+* TINKERPOP3-922 Add a book for Developer Documentation
+* TINKERPOP3-923 Add a book for Tutorials
+* TINKERPOP3-925 Use persisted SparkContext to persist an RDD across Spark jobs.
+* TINKERPOP3-931 Make it possible to extend the core OpProcessor implementations
+* TINKERPOP3-933 Improve release process to get files named properly
+* TINKERPOP3-935 Add missing "close" operation to the session opProcessor docs
+
 TinkerPop 3.0.0 (A Gremlin Rāga in 7/16 Time)
 ---------------------------------------------
 
 image::https://raw.githubusercontent.com/apache/incubator-tinkerpop/master/docs/static/images/gremlin-hindu.png[width=225]
 
+[[release-3-0-2-incubating]]
 TinkerPop 3.0.2 (Release Date: October 19, 2015)
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
@@ -139,6 +225,7 @@ Improvements
 * [TINKERPOP3-853] - TinkerPop Logo in JavaDoc index.html
 * [TINKERPOP3-858] - Cleanup after failed :install
 
+[[release-3-0-1-incubating]]
 TinkerPop 3.0.1 (Release Date: September 2, 2015)
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
diff --git a/docs/src/tutorials-getting-started.asciidoc b/docs/src/tutorials-getting-started.asciidoc
index ccdfb702b61..8560cb72d71 100644
--- a/docs/src/tutorials-getting-started.asciidoc
+++ b/docs/src/tutorials-getting-started.asciidoc
@@ -195,7 +195,7 @@ the console. You would normally refer to them as `T.id` and `T.label`.
 NOTE: The fully qualified name for `T` is `org.apache.tinkerpop.gremlin.structure.T`.
 
 Second, don't forget that you are working with TinkerGraph which allows for identifier assignment. That is _not_ the
-case with most graph databases. For example, don't bother to try with Neo4j or Titan.
+case with most graph databases.
 
 Finally, the label for an `Edge` is required and is thus part of the method signature of `addEdge()`. It is the first
 parameter supplied, followed by the `Vertex` to which `v1` should be connected. Therefore, this usage of `addEdge` is
@@ -561,4 +561,4 @@ Conclusion
 
 ...and that is the end of _The TinkerPop Workout - by Gremlin_. You are hopefully feeling more confident in your
 TinkerPop skills and have a good overview of what the stack has to offer, as well as some entry points to further
-research within the reference documentation. Welcome to The TinkerPop!
\ No newline at end of file
+research within the reference documentation. Welcome to The TinkerPop!
diff --git a/docs/src/upgrade-release-3.1.x-incubating.asciidoc b/docs/src/upgrade-release-3.1.x-incubating.asciidoc
index 3e07c6cc02f..32d23f09c2b 100644
--- a/docs/src/upgrade-release-3.1.x-incubating.asciidoc
+++ b/docs/src/upgrade-release-3.1.x-incubating.asciidoc
@@ -25,9 +25,9 @@ image::https://raw.githubusercontent.com/apache/incubator-tinkerpop/master/docs/
 TinkerPop 3.1.0
 ---------------
 
-*Release Date: NOT OFFICIALLY RELEASED YET*
+*Release Date: November 16, 2015*
 
-Please see the link:https://github.com/apache/incubator-tinkerpop/blob/3.1.0-incubating/CHANGELOG.asciidoc#XXXXXXXXXXXXXXXXXXXXXXXXXXXX[changelog-(NOT FINAL)] for a complete list of all the modifications that are part of this release.
+Please see the link:https://github.com/apache/incubator-tinkerpop/blob/3.1.0-incubating/CHANGELOG.asciidoc#tinkerpop-310-release-date-november-16-2015[changelog] for a complete list of all the modifications that are part of this release.
 
 Additional upgrade information can be found here:
 
diff --git a/giraph-gremlin/pom.xml b/giraph-gremlin/pom.xml
index 28429f2c470..c2c5569fe2f 100644
--- a/giraph-gremlin/pom.xml
+++ b/giraph-gremlin/pom.xml
@@ -21,7 +21,7 @@ limitations under the License.
     <parent>
         <groupId>org.apache.tinkerpop</groupId>
         <artifactId>tinkerpop</artifactId>
-        <version>3.1.0-SNAPSHOT</version>
+        <version>3.1.0-incubating</version>
     </parent>
     <artifactId>giraph-gremlin</artifactId>
     <name>Apache TinkerPop :: Giraph Gremlin</name>
diff --git a/gremlin-console/pom.xml b/gremlin-console/pom.xml
index 50be97eaf7c..a9fbaab9106 100644
--- a/gremlin-console/pom.xml
+++ b/gremlin-console/pom.xml
@@ -21,7 +21,7 @@ limitations under the License.
     <parent>
         <artifactId>tinkerpop</artifactId>
         <groupId>org.apache.tinkerpop</groupId>
-        <version>3.1.0-SNAPSHOT</version>
+        <version>3.1.0-incubating</version>
     </parent>
     <artifactId>gremlin-console</artifactId>
     <name>Apache TinkerPop :: Gremlin Console</name>
diff --git a/gremlin-core/pom.xml b/gremlin-core/pom.xml
index df287fa076d..32004bad8e9 100644
--- a/gremlin-core/pom.xml
+++ b/gremlin-core/pom.xml
@@ -20,7 +20,7 @@ limitations under the License.
     <parent>
         <groupId>org.apache.tinkerpop</groupId>
         <artifactId>tinkerpop</artifactId>
-        <version>3.1.0-SNAPSHOT</version>
+        <version>3.1.0-incubating</version>
     </parent>
     <artifactId>gremlin-core</artifactId>
     <name>Apache TinkerPop :: Gremlin Core</name>
diff --git a/gremlin-driver/pom.xml b/gremlin-driver/pom.xml
index bdf2af6c24d..ac0e1f1a7bf 100644
--- a/gremlin-driver/pom.xml
+++ b/gremlin-driver/pom.xml
@@ -21,7 +21,7 @@ limitations under the License.
     <parent>
         <groupId>org.apache.tinkerpop</groupId>
         <artifactId>tinkerpop</artifactId>
-        <version>3.1.0-SNAPSHOT</version>
+        <version>3.1.0-incubating</version>
     </parent>
     <artifactId>gremlin-driver</artifactId>
     <name>Apache TinkerPop :: Gremlin Driver</name>
diff --git a/gremlin-groovy-test/pom.xml b/gremlin-groovy-test/pom.xml
index ab3d253d9d7..5f581450a91 100644
--- a/gremlin-groovy-test/pom.xml
+++ b/gremlin-groovy-test/pom.xml
@@ -21,7 +21,7 @@ limitations under the License.
     <parent>
         <groupId>org.apache.tinkerpop</groupId>
         <artifactId>tinkerpop</artifactId>
-        <version>3.1.0-SNAPSHOT</version>
+        <version>3.1.0-incubating</version>
     </parent>
     <artifactId>gremlin-groovy-test</artifactId>
     <name>Apache TinkerPop :: Gremlin Groovy Test</name>
diff --git a/gremlin-groovy/pom.xml b/gremlin-groovy/pom.xml
index eefc871980f..b6f611f3db5 100644
--- a/gremlin-groovy/pom.xml
+++ b/gremlin-groovy/pom.xml
@@ -21,7 +21,7 @@ limitations under the License.
     <parent>
         <groupId>org.apache.tinkerpop</groupId>
         <artifactId>tinkerpop</artifactId>
-        <version>3.1.0-SNAPSHOT</version>
+        <version>3.1.0-incubating</version>
     </parent>
     <artifactId>gremlin-groovy</artifactId>
     <name>Apache TinkerPop :: Gremlin Groovy</name>
diff --git a/gremlin-server/pom.xml b/gremlin-server/pom.xml
index a5d79dddc0a..296809947a5 100644
--- a/gremlin-server/pom.xml
+++ b/gremlin-server/pom.xml
@@ -21,7 +21,7 @@ limitations under the License.
     <parent>
         <groupId>org.apache.tinkerpop</groupId>
         <artifactId>tinkerpop</artifactId>
-        <version>3.1.0-SNAPSHOT</version>
+        <version>3.1.0-incubating</version>
     </parent>
     <artifactId>gremlin-server</artifactId>
     <name>Apache TinkerPop :: Gremlin Server</name>
diff --git a/gremlin-shaded/pom.xml b/gremlin-shaded/pom.xml
index 8874cae0d12..f45b1ca6893 100644
--- a/gremlin-shaded/pom.xml
+++ b/gremlin-shaded/pom.xml
@@ -20,7 +20,7 @@ limitations under the License.
     <parent>
         <groupId>org.apache.tinkerpop</groupId>
         <artifactId>tinkerpop</artifactId>
-        <version>3.1.0-SNAPSHOT</version>
+        <version>3.1.0-incubating</version>
     </parent>
     <artifactId>gremlin-shaded</artifactId>
     <name>Apache TinkerPop :: Gremlin Shaded</name>
diff --git a/gremlin-test/pom.xml b/gremlin-test/pom.xml
index 4433c14bbc5..fa1f88bae48 100644
--- a/gremlin-test/pom.xml
+++ b/gremlin-test/pom.xml
@@ -21,7 +21,7 @@ limitations under the License.
     <parent>
         <groupId>org.apache.tinkerpop</groupId>
         <artifactId>tinkerpop</artifactId>
-        <version>3.1.0-SNAPSHOT</version>
+        <version>3.1.0-incubating</version>
     </parent>
     <artifactId>gremlin-test</artifactId>
     <name>Apache TinkerPop :: Gremlin Test</name>
diff --git a/hadoop-gremlin/pom.xml b/hadoop-gremlin/pom.xml
index 26b0b50026f..71f2b82d014 100644
--- a/hadoop-gremlin/pom.xml
+++ b/hadoop-gremlin/pom.xml
@@ -21,7 +21,7 @@ limitations under the License.
     <parent>
         <groupId>org.apache.tinkerpop</groupId>
         <artifactId>tinkerpop</artifactId>
-        <version>3.1.0-SNAPSHOT</version>
+        <version>3.1.0-incubating</version>
     </parent>
     <artifactId>hadoop-gremlin</artifactId>
     <name>Apache TinkerPop :: Hadoop Gremlin</name>
diff --git a/neo4j-gremlin/pom.xml b/neo4j-gremlin/pom.xml
index eca07d1c8ee..50001023dbd 100644
--- a/neo4j-gremlin/pom.xml
+++ b/neo4j-gremlin/pom.xml
@@ -21,7 +21,7 @@ limitations under the License.
     <parent>
         <groupId>org.apache.tinkerpop</groupId>
         <artifactId>tinkerpop</artifactId>
-        <version>3.1.0-SNAPSHOT</version>
+        <version>3.1.0-incubating</version>
     </parent>
     <artifactId>neo4j-gremlin</artifactId>
     <name>Apache TinkerPop :: Neo4j Gremlin</name>
diff --git a/pom.xml b/pom.xml
index f569cefbd9c..a59fba5d005 100644
--- a/pom.xml
+++ b/pom.xml
@@ -25,7 +25,7 @@ limitations under the License.
     </parent>
     <groupId>org.apache.tinkerpop</groupId>
     <artifactId>tinkerpop</artifactId>
-    <version>3.1.0-SNAPSHOT</version>
+    <version>3.1.0-incubating</version>
     <packaging>pom</packaging>
     <name>Apache TinkerPop</name>
     <description>A Graph Computing Framework</description>
diff --git a/spark-gremlin/pom.xml b/spark-gremlin/pom.xml
index 1f4c546667e..396c4e61941 100644
--- a/spark-gremlin/pom.xml
+++ b/spark-gremlin/pom.xml
@@ -24,7 +24,7 @@
     <parent>
         <groupId>org.apache.tinkerpop</groupId>
         <artifactId>tinkerpop</artifactId>
-        <version>3.1.0-SNAPSHOT</version>
+        <version>3.1.0-incubating</version>
     </parent>
     <artifactId>spark-gremlin</artifactId>
     <name>Apache TinkerPop :: Spark Gremlin</name>
diff --git a/tinkergraph-gremlin/pom.xml b/tinkergraph-gremlin/pom.xml
index 8676a9de386..04d5832e250 100644
--- a/tinkergraph-gremlin/pom.xml
+++ b/tinkergraph-gremlin/pom.xml
@@ -21,7 +21,7 @@ limitations under the License.
     <parent>
         <groupId>org.apache.tinkerpop</groupId>
         <artifactId>tinkerpop</artifactId>
-        <version>3.1.0-SNAPSHOT</version>
+        <version>3.1.0-incubating</version>
     </parent>
     <artifactId>tinkergraph-gremlin</artifactId>
     <name>Apache TinkerPop :: TinkerGraph Gremlin</name>
