diff --git a/examples/java/build.gradle b/examples/java/build.gradle
index 40c36c947a35..c4d045d49fd7 100644
--- a/examples/java/build.gradle
+++ b/examples/java/build.gradle
@@ -84,7 +84,6 @@ dependencies {
   compile "org.apache.httpcomponents:httpcore:4.4.13"
   compile library.java.google_oauth_client
   compile library.java.vendored_grpc_1_26_0
-  //compile "com.github.spotbugs:spotbugs-annotations:3.1.9"
   runtime project(path: ":runners:direct-java", configuration: "shadow")
   testCompile project(":sdks:java:io:google-cloud-platform")
   testCompile project(":sdks:java:extensions:ml")
diff --git a/runners/flink/1.11/build.gradle b/runners/flink/1.11/build.gradle
index 2aaa178d485b..b1b2e110de07 100644
--- a/runners/flink/1.11/build.gradle
+++ b/runners/flink/1.11/build.gradle
@@ -17,7 +17,6 @@
  */
 
 def basePath = '..'
-
 /* All properties required for loading the Flink build script */
 project.ext {
   // Set the version of all Flink-related dependencies here.
diff --git a/runners/flink/1.12/build.gradle b/runners/flink/1.12/build.gradle
index be76ac9d860a..5bfa7e3f38c0 100644
--- a/runners/flink/1.12/build.gradle
+++ b/runners/flink/1.12/build.gradle
@@ -17,7 +17,6 @@
  */
 
 def basePath = '..'
-
 /* All properties required for loading the Flink build script */
 project.ext {
   // Set the version of all Flink-related dependencies here.
diff --git a/runners/flink/1.8/build.gradle b/runners/flink/1.8/build.gradle
index 5489aec11194..86edb0e20251 100644
--- a/runners/flink/1.8/build.gradle
+++ b/runners/flink/1.8/build.gradle
@@ -17,7 +17,6 @@
  */
 
 def basePath = '..'
-
 /* All properties required for loading the Flink build script */
 project.ext {
   // Set the version of all Flink-related dependencies here.
diff --git a/runners/flink/1.9/build.gradle b/runners/flink/1.9/build.gradle
index a94187d7655a..0c2c01253e78 100644
--- a/runners/flink/1.9/build.gradle
+++ b/runners/flink/1.9/build.gradle
@@ -17,7 +17,6 @@
  */
 
 def basePath = '..'
-
 /* All properties required for loading the Flink build script */
 project.ext {
   // Set the version of all Flink-related dependencies here.
@@ -32,4 +31,3 @@ project.ext {
 
 // Load the main build script which contains all build logic.
 apply from: "$basePath/flink_runner.gradle"
-
diff --git a/runners/google-cloud-dataflow-java/worker/build.gradle b/runners/google-cloud-dataflow-java/worker/build.gradle
index fb3d68a9c324..c28999dca62c 100644
--- a/runners/google-cloud-dataflow-java/worker/build.gradle
+++ b/runners/google-cloud-dataflow-java/worker/build.gradle
@@ -105,6 +105,7 @@ dependencies {
   compile "org.conscrypt:conscrypt-openjdk-uber:2.2.1"
   compile library.java.slf4j_jdk14
   permitUnusedDeclared library.java.slf4j_jdk14
+
   // All test sourceset dependencies can be marked as shadowTest since we create an uber jar without
   // relocating any code.
   shadowTest project(path: ":runners:core-java", configuration: "testRuntime")
diff --git a/runners/google-cloud-dataflow-java/worker/legacy-worker/build.gradle b/runners/google-cloud-dataflow-java/worker/legacy-worker/build.gradle
index 9a8e31a28463..160e677222ec 100644
--- a/runners/google-cloud-dataflow-java/worker/legacy-worker/build.gradle
+++ b/runners/google-cloud-dataflow-java/worker/legacy-worker/build.gradle
@@ -227,6 +227,7 @@ dependencies {
     compile library.java.vendored_grpc_1_26_0
     compile "org.conscrypt:conscrypt-openjdk-uber:1.3.0"
     compile library.java.slf4j_api
+
     // Any test dependency which intersects with our relocation rules above needs to be relocated
     // as well and placed within the testCompile configuration. Otherwise we can place it within
     // the shadowTest configuration.
diff --git a/runners/portability/java/build.gradle b/runners/portability/java/build.gradle
index ec711b745702..b901548c3ae6 100644
--- a/runners/portability/java/build.gradle
+++ b/runners/portability/java/build.gradle
@@ -39,6 +39,7 @@ dependencies {
   compile library.java.slf4j_api
   compile library.java.joda_time
   compile "org.hamcrest:hamcrest:2.1"
+
   testCompile project(path: ":runners:core-construction-java", configuration: "testRuntime")
   testCompile library.java.hamcrest_core
   testCompile library.java.junit
diff --git a/sdks/java/build-tools/build.gradle b/sdks/java/build-tools/build.gradle
index ae0b4aa681d1..9c66c06bfa9b 100644
--- a/sdks/java/build-tools/build.gradle
+++ b/sdks/java/build-tools/build.gradle
@@ -19,5 +19,4 @@
 plugins { id 'org.apache.beam.module' }
 applyJavaNature(exportJavadoc: false, publish: false)
 
-description = "Apache Beam :: SDKs :: Java :: Build Tools"
-
+description = "Apache Beam :: SDKs :: Java :: Build Tools"
\ No newline at end of file
diff --git a/sdks/java/extensions/sql/datacatalog/build.gradle b/sdks/java/extensions/sql/datacatalog/build.gradle
index 5f0a5d747c8b..c551c5dac582 100644
--- a/sdks/java/extensions/sql/datacatalog/build.gradle
+++ b/sdks/java/extensions/sql/datacatalog/build.gradle
@@ -34,6 +34,8 @@ dependencies {
   compile library.java.vendored_guava_26_0_jre
   compile library.java.slf4j_api
   compile project(path: ":sdks:java:core", configuration: "shadow")
+  compile "org.threeten:threetenbp:1.4.5"
+  compile library.java.protobuf_java
   provided project(":sdks:java:extensions:sql")
 
   // Dependencies for the example
@@ -42,6 +44,8 @@ dependencies {
 
   testCompile project(":sdks:java:extensions:sql:zetasql")
   testRuntimeOnly library.java.slf4j_simple
+  permitUnusedDeclared "org.threeten:threetenbp:1.4.5"
+  permitUnusedDeclared library.java.protobuf_java
 }
 
 task runDataCatalogExample(type: JavaExec) {
diff --git a/sdks/java/io/google-cloud-platform/build.gradle b/sdks/java/io/google-cloud-platform/build.gradle
index e5927525b371..1087e7fe7e57 100644
--- a/sdks/java/io/google-cloud-platform/build.gradle
+++ b/sdks/java/io/google-cloud-platform/build.gradle
@@ -61,7 +61,6 @@ dependencies {
   testImplementation 'io.grpc:grpc-core:1.27.2'
   compile library.java.grpc_netty
   compile library.java.grpc_google_cloud_pubsub_v1
-  //compile library.java.grpc_google_cloud_pubsublite_v1
   compile library.java.guava
   compile library.java.jackson_databind
   compile library.java.joda_time
diff --git a/sdks/java/io/hadoop-format/build.gradle b/sdks/java/io/hadoop-format/build.gradle
index 95fc07a0257d..6e724a22bb9a 100644
--- a/sdks/java/io/hadoop-format/build.gradle
+++ b/sdks/java/io/hadoop-format/build.gradle
@@ -69,7 +69,6 @@ dependencies {
   testCompile project(path: ":sdks:java:testing:test-utils", configuration: "testRuntime")
   testCompile project(":sdks:java:io:jdbc")
   testCompile project(path: ":examples:java", configuration: "testRuntime")
-
   testCompile "org.elasticsearch.plugin:transport-netty4-client:$elastic_search_version"
   testCompile library.java.testcontainers_elasticsearch
   testCompile "org.elasticsearch.client:elasticsearch-rest-high-level-client:$elastic_search_version"
diff --git a/sdks/java/io/kafka/build.gradle b/sdks/java/io/kafka/build.gradle
index 6a76ae837fc9..9207a661ab39 100644
--- a/sdks/java/io/kafka/build.gradle
+++ b/sdks/java/io/kafka/build.gradle
@@ -59,7 +59,7 @@ dependencies {
   compile library.java.jackson_databind
   compile "org.springframework:spring-expression:4.3.18.RELEASE"
   compile ("io.confluent:kafka-avro-serializer:5.3.2") {
-    // It depends on "spotbugs-annotations:3.1.9" which clashes with current
+    // zookeeper depends on "spotbugs-annotations:3.1.9" which clashes with current
     // "spotbugs-annotations:3.1.12" used in Beam. Not required.
     exclude group: "org.apache.zookeeper", module: "zookeeper"
     // "kafka-clients" has to be provided since user can use its own version.
diff --git a/sdks/java/io/kinesis/expansion-service/build.gradle b/sdks/java/io/kinesis/expansion-service/build.gradle
index 90cededf0b9e..1e07a04fc531 100644
--- a/sdks/java/io/kinesis/expansion-service/build.gradle
+++ b/sdks/java/io/kinesis/expansion-service/build.gradle
@@ -34,4 +34,3 @@ ext.summary = "Expansion service serving KinesisIO"
 dependencies {
   runtime library.java.slf4j_jdk14
 }
-
diff --git a/sdks/java/testing/kafka-service/build.gradle b/sdks/java/testing/kafka-service/build.gradle
index 6cfc8ab0466e..f0966744eb0c 100644
--- a/sdks/java/testing/kafka-service/build.gradle
+++ b/sdks/java/testing/kafka-service/build.gradle
@@ -13,7 +13,7 @@
  * distributed under the License is distributed on an AS IS BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
- * limitations under the License
+ * limitations under the License.
  */
 import com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar
 
diff --git a/sdks/java/testing/load-tests/build.gradle b/sdks/java/testing/load-tests/build.gradle
index 6e7840bb8947..7780a9786cda 100644
--- a/sdks/java/testing/load-tests/build.gradle
+++ b/sdks/java/testing/load-tests/build.gradle
@@ -61,7 +61,6 @@ configurations {
 
 dependencies {
   compile library.java.kafka_clients
-
   compile project(path: ":sdks:java:core", configuration: "shadow")
   compile project(":sdks:java:io:synthetic")
   compile project(":sdks:java:testing:test-utils")
@@ -74,6 +73,7 @@ dependencies {
   compile library.java.vendored_guava_26_0_jre
   compile "org.checkerframework:checker-qual:2.0.0"
   compile library.java.slf4j_api
+
   gradleRun project(project.path)
   gradleRun project(path: runnerDependency, configuration: runnerConfiguration)
 
diff --git a/sdks/java/testing/test-utils/build.gradle b/sdks/java/testing/test-utils/build.gradle
index 48dc7e0d244b..0217e6dbd6e4 100644
--- a/sdks/java/testing/test-utils/build.gradle
+++ b/sdks/java/testing/test-utils/build.gradle
@@ -37,6 +37,7 @@ dependencies {
   compile "org.apache.httpcomponents:httpclient:4.5.10"
   compile "org.apache.httpcomponents:httpcore:4.4.12"
   compile library.java.slf4j_api
+
   testCompile library.java.junit
   testRuntimeOnly project(path: ":runners:direct-java", configuration: "shadowTest")
 }
