diff --git a/core/sql/common/ComViewColUsage.h b/core/sql/common/ComViewColUsage.h
index 7d061c7219..1ec12c6334 100644
--- a/core/sql/common/ComViewColUsage.h
+++ b/core/sql/common/ComViewColUsage.h
@@ -88,9 +88,9 @@ class ComViewColUsage
 
   void packUsage (NAString &viewColUsageStr)
   {
-      // usage contains 2 int64 and 3 int32, 100 chars is big enough to hold 
+      // usage contains 2 int64 and 3 int32, 200 chars is big enough to hold 
       // the string representation
-      char buf[100];
+      char buf[200];
       str_sprintf(buf, "viewUID: %Ld viewCol: %d refUID: %Ld refCol: %d refType: %d;",
                   viewUID_, viewColNumber_,
                   refdUID_, refdColNumber_, refdObjectType_);
diff --git a/core/sql/sqlcomp/CmpSeabaseDDLview.cpp b/core/sql/sqlcomp/CmpSeabaseDDLview.cpp
index aaeb0d307e..31bf68a70d 100644
--- a/core/sql/sqlcomp/CmpSeabaseDDLview.cpp
+++ b/core/sql/sqlcomp/CmpSeabaseDDLview.cpp
@@ -265,7 +265,7 @@ short CmpSeabaseDDL::buildViewTblColUsage(const StmtDDLCreateView * createViewPa
      int32_t usingColNum = vctcu.getUsingViewColumnNumber();
 
      // Get column name for view
-     const ComTdbVirtTableColumnInfo * colInfo = &colInfoArray[usingColNum];
+     //const ComTdbVirtTableColumnInfo * colInfo = &colInfoArray[usingColNum];
 
      // Get column number for referenced table
      const ColRefName &usedColRef = vctcu.getUsedObjectColumnName();
@@ -998,11 +998,14 @@ void CmpSeabaseDDL::createSeabaseView(
       return;
     }
 
-  if (updateTextTable(&cliInterface, objUID, COM_VIEW_REF_COLS_TEXT, 0, viewColUsageText))
+  if (!viewColUsageText.isNull())
     {
-      deallocEHI(ehi);
-      processReturn();
-      return;
+      if (updateTextTable(&cliInterface, objUID, COM_VIEW_REF_COLS_TEXT, 0, viewColUsageText))
+        {
+          deallocEHI(ehi);
+          processReturn();
+          return;
+        }
     }
 
   if (updateViewUsage(createViewNode, objUID, &cliInterface))
diff --git a/core/sql/sqlcomp/PrivMgrPrivileges.cpp b/core/sql/sqlcomp/PrivMgrPrivileges.cpp
index c3b116aab1..a1af7cbad1 100644
--- a/core/sql/sqlcomp/PrivMgrPrivileges.cpp
+++ b/core/sql/sqlcomp/PrivMgrPrivileges.cpp
@@ -1127,6 +1127,22 @@ PrivStatus PrivMgrPrivileges::grantColumnPrivileges(
     }
   }
 
+  // Verify that view-col <=> referenced_col relationship exists
+  if (objectType == COM_VIEW_OBJECT)
+  {
+    ViewUsage myUsage;
+    myUsage.viewUID = objectUID_;
+    PrivMgrMDAdmin admin(trafMetadataLocation_, metadataLocation_, pDiags_);
+    if (admin.getViewColUsages(myUsage) == STATUS_ERROR)
+      return STATUS_ERROR;
+    if (myUsage.viewColUsagesStr.empty())
+    {
+      *pDiags_ << DgSqlCode (-CAT_COLUMN_PRIVILEGE_NOT_ALLOWED)
+               << DgTableName (objectName_.c_str());
+       return STATUS_ERROR;
+    }
+  }
+
   // Walk the list of column privileges to grant, and either insert a new
   // row in the COLUMN_PRIVILEGES table or update an existing row.  
 
